---
title: "curatedMetagenomicData"
author: "NSharma"
date: "26/09/2019"
output: html_document
---

```{r setup, include=FALSE}


#BiocManager::install("curatedMetagenomicData")
#BiocManager::install("ExperimentHub")
library(ExperimentHub)
library(curatedMetagenomicData)
suppressPackageStartupMessages(library(ExperimentHub))
library(phyloseq)
library(microbiome)
library(forcats)


eh = ExperimentHub()
myquery = query(eh, "curatedMetagenomicData")

?combined_metadata
View(combined_metadata)
table(combined_metadata$disease)

# Get the stool samples
esl <- curatedMetagenomicData("*metaphlan_bugs_list.stool*", dryrun = FALSE)

# merge
eset <- mergeData(esl)

# the relab=FALSE is essential as it Absolute Raw Count Data
pseq = ExpressionSet2phyloseq(eset,relab=FALSE ,simplify = TRUE, phylogenetictree = TRUE)

#For the MetaPhlAn2 bugs datasets (but not other data types), you gain a lot of taxonomy-aware, ecological analysis and plotting by conversion to a phyloseq class object. curatedMetagenomicData provides the ExpressionSet2phyloseq() function to make this easy:





```

## R Markdown

```{r}
# Create mnd variable
# grabs the nationality from phyloseq
mnd.country <- get_variable(pseq, "country" )
table(mnd.country)

# this collapses the two varialbes into a new variable AB in this case. The command for multiple changes is fct_collapse(x, AB = c("A","B"), DE = c("D","E"))
# Not used BGD Bangladesh , CAN Canada, CHN, China. DNK denmark, FJI Figi, HUN Hungry, ISL isle of man , MDG, MNG mongolia, PER Iran, RUS RUSSIA, 

mnd.country <- fct_collapse(mnd.country, LOW = c("NOR","SWE","FIN") , MEDIUM = c("ITA","AUT","DEU","GBR","ESP","FRA","LUX","NLD"), HIGH = c("USA") )

#mnd.country <- fct_collapse(mnd.country, LOW = c("Scandinavia","EasternEurope") , MEDIUM = c("SouthEurope","CentralEurope", "UKIE"), HIGH = c("US") )

# reorder
mnd.country <- factor(mnd.country, levels = (c("LOW", "MEDIUM", "HIGH")))
levels(mnd.country)

# creates a new variable in the phyloseq called mnd
sample_data(pseq)$mnd = mnd.country

# checks that it has worked.
get_variable(pseq, "mnd")



```

```{r MND table summary}

table(sample_data(pseq.subset)$mnd)

table(sample_data(pseq.subset)$country)

```
## Check OTUs/Phylums
Prune(remove) taxa because we want to remove OTUs/taxa that aren't in any of the samples
Don't need to prune this taxa because OTUs are found in all samples.
Check this everytime you use the subset_samples function because of the pseq object being filtered. 
```{r check OTUs}
# Check if any OTUs are not present in any samples
any(taxa_sums(pseq.subset) == 0)
# FALSE, therefore, OTUs are found in all samples and we don't need to prune the taxa. If the answer is TRUE, some OTUs are not present in any samples. This is usually the case when data is subset to remove some samples. OTUs unique to those sample are not removed along with the samples. Therefore, it is important to check this everytime the phyloseq object is filtered for samples using `subset_samples` function.

head(meta(pseq))

```

## Prepare Data for Alpha Diversity Visualisation
```{r prepare data for alpha diversity visualisation}

# https://bioconductor.riken.jp/packages/3.5/data/experiment/vignettes/curatedMetagenomicData/inst/doc/curatedMetagenomicData.html
# Prepare data for vizualisation

# Remove   everything without a country
pseq.subset <- subset_samples(pseq, !is.na(mnd))
pseq.subset <- subset_samples(pseq, disease == "healthy")

keepotu = genefilter_sample(pseq, filterfun_sample(topp(0.05)), A=5)
f.pseq <- subset_taxa(pseq, keepotu)

any(taxa_sums(pseq.c) == 0)

genus.pseq <- subset_taxa( pseq.subset, !is.na(Genus))
phylum.pseq <- subset_taxa( pseq, is.na(Class) & !is.na(Phylum))


taxonomy <- tax_table(genus.pseq)
# Pick the core (>0.1% relative abundance in >50% of the samples)
pseq.subsetc <- core(pseq.subset, detection = 0.1/100, prevalence = 50/100)

map_levels(NULL, "Phylum","Genus",pseq.c)

diversities(phylum.pseq, index = "all")



png("alpha.png")

tab <- diversities(pseq.c, index = "all")
stargazer(tab, type = "text",summary = NULL)

tab <- diversities(pseq.subsetc, index = "Shannon")
tab <- diversities(phylum.pseq, index = "all")
stargazer(tab, type = "text",summary = NULL)
kable(head(tab))
plot(tab)
dev.off()

# Get metadata
pseq.meta <- meta(phylum.pseq)
kable(head(pseq.meta))
stargazer(pseq.meta, type = "text",summary = NULL)

# Add the diversity table to metadata
pseq.meta$Shannon <- tab$shannon 
pseq.meta$InverseSimpson <- tab$inverse_simpson
pseq.meta$fisher <- tab$fisher
pseq.meta$chao1 <- tab$chao1
#wt = UniFrac(pseq.csubset, weighted=TRUE, normalized=FALSE, 
#             parallel=FALSE, fast=TRUE)
#plot(hclust(wt), main="Weighted UniFrac distances")

library(ggpubr)

p1 <- ggviolin(pseq.meta, x = "mnd", y = "Shannon",
 add = "boxplot", fill = "mnd", palette = "paired") + theme(axis.text.x = element_text(angle = 90)) + theme(legend.position="none") + xlab("")
  print(p1)
ggsave("nation_violin_shannon.png")






keepotu = genefilter_sample(pseq, filterfun_sample(topp(0.05)), A=5)
summary(keepotu)

fpseq <- subset_taxa(pseq, keepotu)

# This doesnt run. 
# plot_heatmap(fpseq, method="PCoA", distance="bray") 

#keepotu = genefilter_sample(pseq, filterfun_sample(topp(0.05)), A=5)
#summary(keepotu)

# subset_taxa(loman.pseq, keepotu)
#lfilt <- subset_taxa(pseq, keepotu & !is.na(Strain))
#plot_heatmap(lfilt, method="PCoA", distance="bray")

#lfilt <- subset_taxa(pseq, is.na(Class) & !is.na(Phylum))

#any(taxa_sums(lfilt) == 0)


# Alpha diversity again
png("alpha.png")
  
barplot(sort(taxa_sums(pseq.csubset), TRUE)[1:20] / nsamples(pseq.csubset),
        ylab = "Total counts", las = 2)

tab <- diversities(pseq.csubset)

 
tab <- diversities(pseq.subset, index = "all")
stargazer(tab, type = "text",summary = NULL)
kable(head(tab))
plot(tab)
dev.off()

# Get metadata
pseq.meta <- meta(pseq)
kable(head(pseq.meta))
stargazer(pseq.meta, type = "text",summary = NULL)

# Add the diversity table to metadata
pseq.meta$Shannon <- tab$shannon 
pseq.meta$InverseSimpson <- tab$inverse_simpson
pseq.meta$fisher <- tab$fisher
pseq.meta$chao1 <- tab$chao1

```

