---
title: "MRD_final_microbiome"
author: "Reis-Dehabadi, M"
date: "4/10/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = '~/globalmicrobiome')


```



## Data Set-up
```{r Data set-up}

#setwd("/Users/maryam/Scripts/Microbiome") 

	## Get data
	library(foreign)
	library(stargazer)
	library(psych)
	library(plyr)
	library(xtable)
	library(car)
	#library(dplyr)
	library(scales)
	library(reshape) 
	library(gmodels)
	library(nnet)
	library(tidyr)
	library(sjmisc)
	library(RColorBrewer)
  library(htmltools)
	library(broom)

  #library(hrbrthemes)
library(gcookbook)
library(tidyverse)
 #library(gdtools)

#The phyloseq package seeks to address issues with multiple microbiome analysis packages by providing a set of functions that internally manage the organizing, linking, storing, and analyzing of phylogenetic sequencing data. In general, this package is used for UniFrac analyses.
library(phyloseq)
library(microbiome)
library(knitr)
library(ggpubr)
#library(phylogeo)
library(dplyr)
#Graphing package used in phyloseq. To edit the default setting of a plot, you need to use functions in this package.
library(ggplot2)
library(viridis) 
library(cowplot)
# install.packages("devtools")
library(devtools)
# install_github("opisthokonta/tsnemicrobiota")
library(tsnemicrobiota)
# check you've installed this library
library(forcats)

# Data from http://dx.doi.org/10.1038/ncomms7342
#This data set contains genus-level microbiota profiling with HITChip for 1006 western adults with no reported health complications. 



mb.gather <- function (x) {
    # Creates an empty data frame
    mb.group <- data.frame("Location" = c("Central Europe","Eastern Europe","Scandinavia", "Southern Europe","UKIE","USA"), "Mean" = c(1:6)) 
     # groups the countries to match the microbiome classification 
  c.europe <- filter(x,Location=="Belgium"|Location=="Denmark"|Location=="Germany"|Location=="The Netherlands")
    e.europe <- filter(x,Location=="Poland")
    scan <- filter(x,Location=="Finland"|Location=="Norway"|Location=="Sweden")
    s.europe <- filter(x,Location=="France"|Location=="Italy"|Location=="Serbia"|Location=="Spain")
    ukie <- filter(x,Location=="United Kingdom"|Location=="Ireland")
    usa <- filter(x,Location=="United States")
    # gmeans the data and put it into the data frame
    mb.group[1,2] <- mean(c.europe$Value)
    mb.group[2,2] <- mean(e.europe$Value)
    mb.group[3,2] <- mean(scan$Value)
    mb.group[4,2] <- mean(s.europe$Value)
    mb.group[5,2] <- mean(ukie$Value)
    mb.group[6,2] <- mean(usa$Value)
    return(mb.group)
    }

setwd("~/globalmicrobiome")
# files nwed to be in the owrking directory

# For MND
mnd.incidence <- as.data.frame(read_csv("./rawdata/MND-I-B.csv",col_names = TRUE))
mnd.prev <- as.data.frame(read_csv("./rawdata/MND-P-B.csv",col_names = TRUE))

mnd.roc.i <- as.data.frame(read_csv("./rawdata/MND-I-B-RoC.csv",col_names = TRUE))

mnd.roc.p <- as.data.frame(read_csv("./rawdata/MND-P-B-RoC.csv",col_names = TRUE))

# Run function and change name
mnd.incidence <- mb.gather(mnd.incidence)
colnames(mnd.incidence)[2] <- "incidence"

mnd.roc.i <- mb.gather(mnd.roc.i)
colnames(mnd.roc.i)[2] <- "ROC incidence"

mnd.prev <- mb.gather(mnd.prev)
colnames(mnd.prev)[2] <- "prevalance"

mnd.roc.p <- mb.gather(mnd.roc.p)
colnames(mnd.roc.p)[2] <- "ROC prevalance"
# Join them together. 
mnd.ip <- right_join(mnd.prev, mnd.roc.p, by = "Location")
mnd.ip <- right_join(mnd.ip, mnd.incidence, by = "Location")
mnd.ip <- right_join(mnd.ip, mnd.roc.i, by = "Location")

# For IPD

ipd.incidence <- as.data.frame(read_csv("./rawdata/PD-I-B.csv",col_names = TRUE))
ipd.prev <- as.data.frame(read_csv("./rawdata/PD-P-B.csv",col_names = TRUE))

ipd.roc.i <- as.data.frame(read_csv("./rawdata/PD-I-B-RoC.csv",col_names = TRUE))

ipd.roc.p <- as.data.frame(read_csv("./rawdata/PD-P-B-RoC.csv",col_names = TRUE))

# Run function and change name
ipd.incidence <- mb.gather(ipd.incidence)
colnames(ipd.incidence)[2] <- "incidence"

ipd.roc.i <- mb.gather(ipd.roc.i)
colnames(ipd.roc.i)[2] <- "ROC incidence"

ipd.prev <- mb.gather(ipd.prev)
colnames(ipd.prev)[2] <- "prevalance"

ipd.roc.p <- mb.gather(ipd.roc.p)
colnames(ipd.roc.p)[2] <- "ROC prevalance"
# Join them together. 
ipd.ip <- right_join(ipd.prev, ipd.roc.p, by = "Location")
ipd.ip <- right_join(ipd.ip, ipd.incidence, by = "Location")
ipd.ip <- right_join(ipd.ip, ipd.roc.i, by = "Location")


# For AD
ad.incidence <- as.data.frame(read_csv("./rawdata/AD-I-B.csv",col_names = TRUE))
ad.prev <- as.data.frame(read_csv("./rawdata/AD-P-B.csv",col_names = TRUE))

ad.roc.i <- as.data.frame(read_csv("./rawdata/AD-I-B-RoC.csv",col_names = TRUE))

ad.roc.p <- as.data.frame(read_csv("./rawdata/AD-P-B-RoC.csv",col_names = TRUE))

# Run function and change name
ad.incidence <- mb.gather(ad.incidence)
colnames(ad.incidence)[2] <- "incidence"

ad.roc.i <- mb.gather(ad.roc.i)
colnames(ad.roc.i)[2] <- "ROC incidence"

ad.prev <- mb.gather(ad.prev)
colnames(ad.prev)[2] <- "prevalance"

ad.roc.p <- mb.gather(ad.roc.p)
colnames(ad.roc.p)[2] <- "ROC prevalance"
# Join them together. 
ad.ip <- right_join(ad.prev, ad.roc.p, by = "Location")
ad.ip <- right_join(ad.ip, ad.incidence, by = "Location")
ad.ip <- right_join(ad.ip, ad.roc.i, by = "Location")



plot(ad.ip$incidence,ad.ip$prevalance)
plot(ipd.ip$incidence,ipd.ip$prevalance)
plot(mnd.ip$incidence,mnd.ip$prevalance)

```

