---
title: "MRD_final_microbiome"
author: "Reis-Dehabadi, M"
date: "4/10/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = '~/globalmicrobiome')


```



## Data Set-up
```{r Data set-up}

#setwd("/Users/maryam/Scripts/Microbiome") 

	## Get data
	library(foreign)
	library(stargazer)
	library(psych)
	library(plyr)
	library(xtable)
	library(car)
	#library(dplyr)
	library(scales)
	library(reshape) 
	library(gmodels)
	library(nnet)
	library(tidyr)
	library(sjmisc)
	library(RColorBrewer)
  library(htmltools)
	library(broom)

  #library(hrbrthemes)
library(gcookbook)
library(tidyverse)
 #library(gdtools)

#The phyloseq package seeks to address issues with multiple microbiome analysis packages by providing a set of functions that internally manage the organizing, linking, storing, and analyzing of phylogenetic sequencing data. In general, this package is used for UniFrac analyses.
library(phyloseq)
library(microbiome)
library(knitr)
library(ggpubr)
#library(phylogeo)
library(dplyr)
#Graphing package used in phyloseq. To edit the default setting of a plot, you need to use functions in this package.
library(ggplot2)
library(viridis) 
library(cowplot)
# install.packages("devtools")
library(devtools)
# install_github("opisthokonta/tsnemicrobiota")
library(tsnemicrobiota)
# check you've installed this library
library(forcats)

# Data from http://dx.doi.org/10.1038/ncomms7342
#This data set contains genus-level microbiota profiling with HITChip for 1006 western adults with no reported health complications. 



mb.gather <- function (x) {
    # Creates an empty data frame
    mb.group <- data.frame("Location" = c("Central Europe","Eastern Europe","Scandinavia", "Southern Europe","UKIE","USA"), "Mean" = c(1:6)) 
     # groups the countries to match the microbiome classification 
  c.europe <- filter(x,Location=="Belgium"|Location=="Denmark"|Location=="Germany"|Location=="The Netherlands")
    e.europe <- filter(x,Location=="Poland")
    scan <- filter(x,Location=="Finland"|Location=="Norway"|Location=="Sweden")
    s.europe <- filter(x,Location=="France"|Location=="Italy"|Location=="Serbia"|Location=="Spain")
    ukie <- filter(x,Location=="United Kingdom"|Location=="Ireland")
    usa <- filter(x,Location=="United States")
    # gmeans the data and put it into the data frame
    mb.group[1,2] <- mean(c.europe$Value)
    mb.group[2,2] <- mean(e.europe$Value)
    mb.group[3,2] <- mean(scan$Value)
    mb.group[4,2] <- mean(s.europe$Value)
    mb.group[5,2] <- mean(ukie$Value)
    mb.group[6,2] <- mean(usa$Value)
    return(mb.group)
    }

setwd("~/globalmicrobiome")
# files nwed to be in the owrking directory

# For MND
mnd.incidence <- as.data.frame(read_csv("./rawdata/MND-I-B.csv",col_names = TRUE))

mnd.i.12 <- as.data.frame(read_csv("./rawdata/MND-I-B-12.csv",col_names = TRUE))
mnd.i.13 <- as.data.frame(read_csv("./rawdata/MND-I-B-13.csv",col_names = TRUE))
mnd.i.14 <- as.data.frame(read_csv("./rawdata/MND-I-B-14.csv",col_names = TRUE))
mnd.i.15 <- as.data.frame(read_csv("./rawdata/MND-I-B-15.csv",col_names = TRUE))
mnd.i.16 <- as.data.frame(read_csv("./rawdata/MND-I-B-16.csv",col_names = TRUE))


mnd.prev <- as.data.frame(read_csv("./rawdata/MND-P-B.csv",col_names = TRUE))
mnd.p.12 <- as.data.frame(read_csv("./rawdata/MND-P-B-12.csv",col_names = TRUE))
mnd.p.13 <- as.data.frame(read_csv("./rawdata/MND-P-B-13.csv",col_names = TRUE))
mnd.p.14 <- as.data.frame(read_csv("./rawdata/MND-P-B-14.csv",col_names = TRUE))
mnd.p.15 <- as.data.frame(read_csv("./rawdata/MND-P-B-15.csv",col_names = TRUE))
mnd.p.16 <- as.data.frame(read_csv("./rawdata/MND-P-B-16.csv",col_names = TRUE))


mnd.roc.i <- as.data.frame(read_csv("./rawdata/MND-I-B-RoC.csv",col_names = TRUE))

mnd.roc.p <- as.data.frame(read_csv("./rawdata/MND-P-B-RoC.csv",col_names = TRUE))

# Run function and change name
mnd.incidence <- mb.gather(mnd.incidence)
colnames(mnd.incidence)[2] <- 2017

mnd.i.12 <- mb.gather(mnd.i.12)
colnames(mnd.i.12)[2] <- 2012
mnd.i.13 <- mb.gather(mnd.i.13)
colnames(mnd.i.13)[2] <- 2013
mnd.i.14 <- mb.gather(mnd.i.14)
colnames(mnd.i.14)[2] <- 2014
mnd.i.15 <- mb.gather(mnd.i.15)
colnames(mnd.i.15)[2] <- 2015
mnd.i.16 <- mb.gather(mnd.i.16)
colnames(mnd.i.16)[2] <- 2016
#mnd.roc.i <- mb.gather(mnd.roc.i)
#colnames(mnd.roc.i)[2] <- "ROC incidence"

mnd.prev <- mb.gather(mnd.prev)
colnames(mnd.prev)[2] <- 2017
mnd.p.12 <- mb.gather(mnd.p.12)
colnames(mnd.p.12)[2] <- 2012
mnd.p.13 <- mb.gather(mnd.p.13)
colnames(mnd.p.13)[2] <- 2013
mnd.p.14 <- mb.gather(mnd.p.14)
colnames(mnd.p.14)[2] <- 2014
mnd.p.15 <- mb.gather(mnd.p.15)
colnames(mnd.p.15)[2] <- 2015
mnd.p.16 <- mb.gather(mnd.p.16)
colnames(mnd.p.16)[2] <- 2016
#mnd.roc.p <- mb.gather(mnd.roc.p)
#colnames(mnd.roc.p)[2] <- "ROC prevalance"

# Join them together. 

mnd.i <- list( mnd.i.12, mnd.i.13,mnd.i.14,mnd.i.15,mnd.i.16,mnd.incidence) %>% reduce(left_join, by = "Location")

mnd.p <- list( mnd.p.12, mnd.p.13,mnd.p.14,mnd.p.15,mnd.p.16,mnd.incidence) %>% reduce(left_join, by = "Location")

# For IPD

ipd.incidence <- as.data.frame(read_csv("./rawdata/PD-I-B.csv",col_names = TRUE))

ipd.i.12 <- as.data.frame(read_csv("./rawdata/PD-I-B-12.csv",col_names = TRUE))
ipd.i.13 <- as.data.frame(read_csv("./rawdata/PD-I-B-13.csv",col_names = TRUE))
ipd.i.14 <- as.data.frame(read_csv("./rawdata/PD-I-B-14.csv",col_names = TRUE))
ipd.i.15 <- as.data.frame(read_csv("./rawdata/PD-I-B-15.csv",col_names = TRUE))
ipd.i.16 <- as.data.frame(read_csv("./rawdata/PD-I-B-16.csv",col_names = TRUE))


ipd.prev <- as.data.frame(read_csv("./rawdata/PD-P-B.csv",col_names = TRUE))
ipd.p.12 <- as.data.frame(read_csv("./rawdata/PD-P-B-12.csv",col_names = TRUE))
ipd.p.13 <- as.data.frame(read_csv("./rawdata/PD-P-B-13.csv",col_names = TRUE))
ipd.p.14 <- as.data.frame(read_csv("./rawdata/PD-P-B-14.csv",col_names = TRUE))
ipd.p.15 <- as.data.frame(read_csv("./rawdata/PD-P-B-15.csv",col_names = TRUE))
ipd.p.16 <- as.data.frame(read_csv("./rawdata/PD-P-B-16.csv",col_names = TRUE))


ipd.roc.i <- as.data.frame(read_csv("./rawdata/PD-I-B-RoC.csv",col_names = TRUE))

ipd.roc.p <- as.data.frame(read_csv("./rawdata/PD-P-B-RoC.csv",col_names = TRUE))

# Run function and change name
ipd.incidence <- mb.gather(ipd.incidence)
colnames(ipd.incidence)[2] <- 2017

ipd.i.12 <- mb.gather(ipd.i.12)
colnames(ipd.i.12)[2] <- 2012
ipd.i.13 <- mb.gather(ipd.i.13)
colnames(ipd.i.13)[2] <- 2013
ipd.i.14 <- mb.gather(ipd.i.14)
colnames(ipd.i.14)[2] <- 2014
ipd.i.15 <- mb.gather(ipd.i.15)
colnames(ipd.i.15)[2] <- 2015
ipd.i.16 <- mb.gather(ipd.i.16)
colnames(ipd.i.16)[2] <- 2016
#ipd.roc.i <- mb.gather(ipd.roc.i)
#colnames(ipd.roc.i)[2] <- "ROC incidence"

ipd.prev <- mb.gather(ipd.prev)
colnames(ipd.prev)[2] <- 2017
ipd.p.12 <- mb.gather(ipd.p.12)
colnames(ipd.p.12)[2] <- 2012
ipd.p.13 <- mb.gather(ipd.p.13)
colnames(ipd.p.13)[2] <- 2013
ipd.p.14 <- mb.gather(ipd.p.14)
colnames(ipd.p.14)[2] <- 2014
ipd.p.15 <- mb.gather(ipd.p.15)
colnames(ipd.p.15)[2] <- 2015
ipd.p.16 <- mb.gather(ipd.p.16)
colnames(ipd.p.16)[2] <- 2016
#ipd.roc.p <- mb.gather(ipd.roc.p)
#colnames(ipd.roc.p)[2] <- "ROC prevalance"

# Join them together. 

ipd.i <- list( ipd.i.12, ipd.i.13,ipd.i.14,ipd.i.15,ipd.i.16,ipd.incidence) %>% reduce(left_join, by = "Location")

ipd.p <- list( ipd.p.12, ipd.p.13,ipd.p.14,ipd.p.15,ipd.p.16,ipd.incidence) %>% reduce(left_join, by = "Location")



# For AD

ad.incidence <- as.data.frame(read_csv("./rawdata/AD-I-B.csv",col_names = TRUE))

ad.i.12 <- as.data.frame(read_csv("./rawdata/AD-I-B-12.csv",col_names = TRUE))
ad.i.13 <- as.data.frame(read_csv("./rawdata/AD-I-B-13.csv",col_names = TRUE))
ad.i.14 <- as.data.frame(read_csv("./rawdata/AD-I-B-14.csv",col_names = TRUE))
ad.i.15 <- as.data.frame(read_csv("./rawdata/AD-I-B-15.csv",col_names = TRUE))
ad.i.16 <- as.data.frame(read_csv("./rawdata/AD-I-B-16.csv",col_names = TRUE))


ad.prev <- as.data.frame(read_csv("./rawdata/AD-P-B.csv",col_names = TRUE))
ad.p.12 <- as.data.frame(read_csv("./rawdata/AD-P-B-12.csv",col_names = TRUE))
ad.p.13 <- as.data.frame(read_csv("./rawdata/AD-P-B-13.csv",col_names = TRUE))
ad.p.14 <- as.data.frame(read_csv("./rawdata/AD-P-B-14.csv",col_names = TRUE))
ad.p.15 <- as.data.frame(read_csv("./rawdata/AD-P-B-15.csv",col_names = TRUE))
ad.p.16 <- as.data.frame(read_csv("./rawdata/AD-P-B-16.csv",col_names = TRUE))


ad.roc.i <- as.data.frame(read_csv("./rawdata/AD-I-B-RoC.csv",col_names = TRUE))

ad.roc.p <- as.data.frame(read_csv("./rawdata/AD-P-B-RoC.csv",col_names = TRUE))

# Run function and change name
ad.incidence <- mb.gather(ad.incidence)
colnames(ad.incidence)[2] <- 2017

ad.i.12 <- mb.gather(ad.i.12)
colnames(ad.i.12)[2] <- 2012
ad.i.13 <- mb.gather(ad.i.13)
colnames(ad.i.13)[2] <- 2013
ad.i.14 <- mb.gather(ad.i.14)
colnames(ad.i.14)[2] <- 2014
ad.i.15 <- mb.gather(ad.i.15)
colnames(ad.i.15)[2] <- 2015
ad.i.16 <- mb.gather(ad.i.16)
colnames(ad.i.16)[2] <- 2016
#ad.roc.i <- mb.gather(ad.roc.i)
#colnames(ad.roc.i)[2] <- "ROC incidence"

ad.prev <- mb.gather(ad.prev)
colnames(ad.prev)[2] <- 2017
ad.p.12 <- mb.gather(ad.p.12)
colnames(ad.p.12)[2] <- 2012
ad.p.13 <- mb.gather(ad.p.13)
colnames(ad.p.13)[2] <- 2013
ad.p.14 <- mb.gather(ad.p.14)
colnames(ad.p.14)[2] <- 2014
ad.p.15 <- mb.gather(ad.p.15)
colnames(ad.p.15)[2] <- 2015
ad.p.16 <- mb.gather(ad.p.16)
colnames(ad.p.16)[2] <- 2016
#ad.roc.p <- mb.gather(ad.roc.p)
#colnames(ad.roc.p)[2] <- "ROC prevalance"

# Join them together. 

ad.i <- list( ad.i.12, ad.i.13,ad.i.14,ad.i.15,ad.i.16,ad.incidence) %>% reduce(left_join, by = "Location")

ad.p <- list( ad.p.12, ad.p.13,ad.p.14,ad.p.15,ad.p.16,ad.incidence) %>% reduce(left_join, by = "Location")

library(reshape)
as.numeric.factor <- function(x) {as.numeric(levels(x))[x]}

require(cowplot)
theme_set(theme_cowplot(font_size=12)) # reduce default font size

mnd.i.melt = melt(mnd.i, id = 'Location')
mnd.i.melt$variable <- as.numeric.factor(mnd.i.melt$variable)
p1 <- ggplot(mnd.i.melt, aes(x = variable, y = value, colour = Location)) + 
  geom_line() + 
  ylab(label="mnd Incidence") + 
  xlab("Year") + theme(legend.position="none")

mnd.p.melt = melt(mnd.i, id = 'Location')
mnd.p.melt$variable <- as.numeric.factor(mnd.p.melt$variable)
p2 <-  ggplot(mnd.p.melt, aes(x = variable, y = value, colour = Location)) + 
  geom_line() + 
  ylab(label="mnd Prevalance") + 
  xlab("Year") + theme(legend.position="none")

ipd.i.melt = melt(ipd.i, id = 'Location')
ipd.i.melt$variable <- as.numeric.factor(ipd.i.melt$variable)
p3 <- ggplot(ipd.i.melt, aes(x = variable, y = value, colour = Location)) + 
  geom_line() + 
  ylab(label="ipd Incidence") + 
  xlab("Year") + theme(legend.position="none")

ipd.p.melt = melt(ipd.i, id = 'Location')
ipd.p.melt$variable <- as.numeric.factor(ipd.p.melt$variable)
p4 <-  ggplot(ipd.p.melt, aes(x = variable, y = value, colour = Location)) + 
  geom_line() + 
  ylab(label="ipd Prevalance") + 
  xlab("Year") + theme(legend.position="none")

ad.i.melt = melt(ad.i, id = 'Location')
ad.i.melt$variable <- as.numeric.factor(ad.i.melt$variable)
p5 <-  ggplot(ad.i.melt, aes(x = variable, y = value, colour = Location)) + 
  geom_line() + 
  ylab(label="AD Incidence") + 
  xlab("Year") + theme(legend.position="none")

ad.p.melt = melt(ad.i, id = 'Location')
ad.p.melt$variable <- as.numeric.factor(ad.p.melt$variable)
p6 <- ggplot(ad.p.melt, aes(x = variable, y = value, colour = Location)) + 
  geom_line() + 
  ylab(label="AD Prevalance") + 
  xlab("Year") + theme(legend.position="none")

final <- plot_grid(p1, p2,p3,p4,p5,p6, labels = c('A', 'B','C', 'D','E', 'F'),ncol = 2)
ggsave("overview.jpg", final) 
```

