---
title: "Untitled"
author: "NSharma"
date: "2019"
output: html_document
---

```{r setup, include=FALSE}

setwd("/Users/Nik/globalmicrobiome")

library(gcookbook)
library(tidyverse)
 #library(gdtools)

#The phyloseq package seeks to address issues with multiple microbiome analysis packages by providing a set of functions that internally manage the organizing, linking, storing, and analyzing of phylogenetic sequencing data. In general, this package is used for UniFrac analyses.
library(phyloseq)
library(microbiome)
library(knitr)
library(ggpubr)
#library(phylogeo)
library(dplyr)
#Graphing package used in phyloseq. To edit the default setting of a plot, you need to use functions in this package.
library(ggplot2)
library(viridis) 
library(cowplot)
# install.packages("devtools")
library(devtools)
# install_github("opisthokonta/tsnemicrobiota")
library(tsnemicrobiota)
# check you've installed this library
library(forcats)
library(IPCO)
library(phylosmith)
library(phyloseq)

data(atlas1006)   
pseq <- baseline(atlas1006)

##Step 1: Extract the required matrix from Phyloseq object
##first convert the OTU level phyloseq object to species/genus level phyloseq object. Depending on the phyloseq version the taxrank or taxlevel parameter would be used.
#taxlevel parameter is probably for older version to the best of my knowledge

#phyloseq_genus <- tax_glom(pseq, taxlevel="Genus")

#taxrank parameter is in the latest version
phyloseq_genus <- tax_glom(pseq, taxrank="Genus")

#If the above step is already done, then you can start from here.
#Convert phyloseq genus/species level object to matrix
genus_table <- as(otu_table(phyloseq_genus), "matrix")

##Step 2: Using IPCO
#load refernce data IPCO_Healthy/IPCO_HMP
library(IPCO)
data(IPCO_Healthy)

#Do normalisation/transform as suggested as an where required. If the phyloseq data is already normalised and doesnot wish to renormalise, then the user doesnot have to follow IPCO normalisation.

#We recommend Hellinger transformation. To be used on count/prop/relative abundance data
ref_genus <- transform_data(IPCO_Healthy$Genus,"relab")
ref_func <- transform_data(IPCO_Healthy$MetaCyc,"proportion")

#ref_func <- filter_functionality(ref_func, "metacyc",IPCO_Healthy$MetaCyc_coverage,threshold=0.4) # Filtering

ref_func <- filter_functionality(ref_func, "metacyc", IPCO_Healthy$MetaCyc_coverage)
##or
#ref_func <- transform_data(IPCO_Healthy$KEGG,"proportion")

#if query genus table is count data
#query_genus <- transform_data(genus_table,"count") 

#Run IPCO
#Note: all samples should be in columns and features (func/taxa) in rows. If not, transpose the matrix
inferred_MetaCyc <- IPCO(ref_func,ref_genus,genus_table)

# Create mnd variable
# grabs the nationality from phyloseq
con.class <- get_variable(pseq, "nationality" )

# check you've installed this library
library(forcats)

# this collapses the two varialbes into a new variable AB in this case. The command for multiple changes is fct_collapse(x, AB = c("A","B"), DE = c("D","E"))
mnd.class <- fct_collapse(con.class, LOW = c("Scandinavia","EasternEurope") , MEDIUM = c("SouthEurope","CentralEurope", "UKIE"), HIGH = c("US") )

# reorder
mnd.class <- factor(mnd.class, levels = (c("LOW", "MEDIUM", "HIGH")))


library(made4)
library(ade4)

overview(inferred_MetaCyc)

```


```{r , include=FALSE}

#p <- overview(inferred_MetaCyc,classvec=mnd.class, labels=c("red", "blue", "yellow"))
 
#Correspondence Analysis
#The function ord simplifies the running of ordination methods such as principal component, correspondence or non-symmetric correspondence analysis. It provides a wrapper
#which can call each of these methods in ade4. To run a correspondence analysis [6] on
#this dataset.


k.coa<- ord(inferred_MetaCyc, type="coa")
summary(k.coa$ord)
p <- plot(k.coa)

p <- plot(k.coa, classvec=mnd.class, arraycol=c("red", "blue", "yellow" ), genecol="grey3")

```

```{r , include=FALSE}
plotarrays(k.coa, arraylabels=mnd.class)

plotgenes(k.coa)

k.coa2<-ord(inferred_MetaCyc, classvec=mnd.class)
p <- plot(k.coa2)

#Classification and Class Prediction using Between Group
#Analysis
#Between Group Analysis (BGA) is a supervised classification method [3]. The basis
#of BGA is to ordinate the groups rather than the individual samples. In tests on two
#microarray gene expression datasets, BGA performed comparably to supervised classification methods, including support vector machines and artifical neural networks [2].
#To train a dataset, use bga, the projection of test data can be assessed using suppl.
#One leave out cross validation can be perfo

k.bga<-bga(inferred_MetaCyc, type="coa", classvec=mnd.class)
between.graph(k.bga, ax=1) # Show the separation on the first axes(ax

#Meta-analysis of microarray gene expression
#Coinertia analysis cia [4] has been successfully applied to the cross-platform comparison
#(meta-analysis) of microarray gene expression datasets [8]. CIA is a multivariate method
#that identifies trends or co-relationships in multiple datasets which contain the same
#samples. That is either the rows or the columns of a matrix must be ”matchable”. CIA
#can be applied to datasets where the number of variables (genes) far exceeds the number
#of samples (arrays) such is the case with microarray analyses. cia calls coinertia in
#the ade4 package. See the CIA vignette for more details on this method.

#By default the variables (genes) are labelled with the rownames of the matrix. Typically these are spot IDs or Affymetrix accession numbers which are not very easy to
#interpret. But these can be easily labeled by your own labels. For example its often
#useful to labels using HUGO gene symbols. We find the Bioconductor annotate and
#annaffy annotation packages are very useful for this. Alternatively we also use biomaRt
#or Resourcerer or the Stanford Source database.
#In this example we provide annotation from the Source database in khan$annotation.
#The gene symbol are in the column khan$annotation$Symbol

#plot(k.bga, genelabels=gene.symbs)    need to define gene;symbs

heatmap(inferred_MetaCyc)

p <- plot_ly(inferred_MetaCyc = volcano, type = "heatmap")

levelplot(inferred_MetaCyc[1:ncol(inferred_MetaCyc),ncol(inferred_MetaCyc):1])

coinertia_resuls <- check_coinertia(ref_func,genus_table)


```

## R Markdown
