---
title: "curatedMetagenomicData"
author: "NSharma"
date: "26/09/2019"
output:
  word_document: default
  html_document: default
---

```{r , include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = '~/globalmicrobiome')


```



## Setup
```{r setup, include=FALSE}


#BiocManager::install("curatedMetagenomicData")
#BiocManager::install("ExperimentHub")
library(ExperimentHub)
library(curatedMetagenomicData)
suppressPackageStartupMessages(library(ExperimentHub))
library(phyloseq)
library(microbiome)
library(forcats)
library(stargazer)
library(ggpubr)
library(tidyverse)
library(cowplot)
library(vegan)
library(broom)
library(tsnemicrobiota)
library(latex2exp)
library(wesanderson)
library(table1)
library(knitr)
library(kableExtra)


eh = ExperimentHub()
myquery = query(eh, "curatedMetagenomicData")

# Get the stool samples
esl <- curatedMetagenomicData("*metaphlan_bugs_list.stool*", dryrun = FALSE)

# merge
eset <- mergeData(esl)

# the relab=FALSE is essential as it Absolute Raw Count Data
pseq = ExpressionSet2phyloseq(eset,relab=FALSE ,simplify = TRUE, phylogenetictree = TRUE)
orig_pseq <- pseq




#For the MetaPhlAn2 bugs datasets (but not other data types), you gain a lot of taxonomy-aware, ecological analysis and plotting by conversion to a phyloseq class object. curatedMetagenomicData provides the ExpressionSet2phyloseq() function to make this easy:
data(atlas1006)
atlas.pseq <- baseline(atlas1006) # only uses the baseline samples. This is n=1006


```


## Incidence of MND
```{r}

# This block uses the raw data to examin  the incidence of MND to ensure the classificaiton is correct. It assumes that the data is in the correct place so the setwd at the top meeds to be accurate. 
setwd('~/globalmicrobiome') 
# For MND  nrows = 196 as thwew are other things after 196
mnd.i.15_49 <- as.data.frame(read.csv("./rawdata/MND_I_15_49.csv", header=TRUE, nrows = 196))

mnd.i.50_69  <- as.data.frame(read.csv("./rawdata/MND_I_50_69.csv", header=TRUE, nrows = 196))
mnd.i.70 <- as.data.frame(read.csv("./rawdata/MND_I_70plus.csv", header=TRUE, nrows = 196))

mnd.i.70.test <- mnd.i.70
# Run function and change name
mnd.i.15_49 <- mb.gather(mnd.i.15_49)
colnames(mnd.i.15_49)[2] <- "15-49"

# Run function and change name
mnd.i.50_69 <- mb.gather(mnd.i.50_69)
colnames(mnd.i.50_69)[2] <- "50-69"

# Run function and change name
mnd.i.70 <- mb.gather(mnd.i.70)
colnames(mnd.i.70)[2] <- "70"
  
  
# Join them together.   
mnd.i <- list( mnd.i.15_49,mnd.i.50_69,mnd.i.70 ) %>% reduce(left_join, by = "Location")


a.mnd.nationality <- fct_collapse(a.mnd.nationality, LOW = c("EasternEurope") , MEDIUM = c("SouthEurope","CentralEurope" ), HIGH = c("US","UKIE","Scandinavia" ))

mnd.i %>%
  kable(col.names = c("","Incidence","CI","Incidence","CI","Incidence","CI")) %>%
  kable_styling(bootstrap_options = c("striped", "hover"),full_width = F,position = "center") %>%
add_header_above(c("Location" = 1, "15-49" = 2, "50-69" = 2, "70+" = 2)) %>%
  add_header_above(c(" ", "Age (years)" = 6)) %>%
  pack_rows("High MND group", 5,6)%>%
  pack_rows("Low MND group", 2, 2) %>%
  pack_rows("Medium MND group",1,3,4) 
  


# Now do the same for the curated data set
mnd.ic.15_49 <- mb.cgather(mnd.i.15_49)
colnames(mnd.ic.15_49)[2] <- "15-49"


# Run function and change name
mnd.ic.50_69 <- mb.cgather(mnd.i.50_69)
colnames(mnd.ic.50_69)[2] <- "50-69"

# Run function and change name
mnd.ic.70 <- mb.cgather(mnd.i.70)
colnames(mnd.ic.70)[2] <- "70"

# Join them together. 

mnd.ic <- list( mnd.ic.15_49,mnd.ic.50_69,mnd.ic.70 ) %>% reduce(left_join, by = "Location") 

mnd.ic <- dplyr::select(mnd.ic, "Location",  "Value.x", "Lower.bound.x", "Upper.bound.x", "Value.y", "Lower.bound.y", "Upper.bound.y", "Value", "Lower.bound", "Upper.bound" )

# Prep the data for the table




mnd.ic %>%
  kable(col.names = c("","Incidence","Upper","Lower","Incidence","Upper","Lower","Incidence","Upper","Lower"),digits=3) %>%
  kable_styling(bootstrap_options = c("striped", "hover"),full_width = F,position = "left") %>%
add_header_above(c("Location" = 1, "15-49" = 3, "50-69" = 3, "70+" = 3)) %>%
  add_header_above(c(" ", "Age (years)" = 9)) %>%
  pack_rows("Group 1", 4, 7) %>%
  pack_rows("Group 1", 4, 7) %>%
  pack_rows("Group 2", 8, 10)


```

## MND Countries
```{r tables}

# THIS FOR THE CURATED DATA SET

# creates a grouping varaible for BMI
n <- length(sample_data(pseq)$BMI)
data <- data.frame(id=1:n)
data$bmi <- sample_data(pseq)$BMI # creates a new variable with BMI in it
# Classification based upon the ATlas data
data$bmi[data$bmi <18.5] <- "underweight"
data$bmi[data$bmi >=45] <- "super obese"
data$bmi[data$bmi >= 18.5 & data$bmi <25] <- "lean"
data$bmi[data$bmi >= 25 & data$bmi <30] <- "overweight"
data$bmi[data$bmi >= 30 & data$bmi <35] <- "obese"
data$bmi[data$bmi >= 35 & data$bmi <40] <- "severe obese"
data$bmi[data$bmi >= 40 & data$bmi <45] <- "morbid obese"
data$bmi <- factor(data$bmi)    

# creates a new variable in the phyloseq called mnd
sample_data(pseq)$bmi_group = data$bmi
# Check that it works
get_variable(pseq, "bmi_group")

# Create mnd variable
#pseq <- orig_pseq
# grabs the nationality from phyloseq
c.mnd.country <- get_variable(pseq, "country" )
table(c.mnd.country)

# this collapses the two varialbes into a new variable AB in this case. The command for multiple changes is fct_collapse(x, AB = c("A","B"), DE = c("D","E"))
# Not used BGD Bangladesh , CAN Canada, CHN, China. DNK denmark, FJI Figi, HUN Hungry, ISL isle of man , MDG, MNG mongolia, PER Iran, RUS RUSSIA, 

#Adjsuted to match exactly the ATLAS data set

#central Europe = Belgium, Demark, Netherlands, Germany
#southern Europe = France, Italy, Serbia, Spain
#UKIE
#Scandinavia = Norway, Sweden Finland
#Eastern Europe = Poland 


c.mnd.atlas <- fct_collapse(c.mnd.country, Scandinavia = c("NOR","SWE","FIN") ,CentralEurope = c("DNK","DEU","NLD"), SouthEurope = c("ITA","ESP","FRA"), US = "USA", UKIE = "GBR" )

#drops unused levels
c.mnd.atlas <- droplevels(c.mnd.atlas,c("ISL","HUN","EST", "SVK", "LUX", "CAN", "AUT", "BGD","FJI","CHN","MDG","MNG","PER","RUS","TZA") ) 

sample_data(pseq)$nationality = c.mnd.atlas


#mnd.country <- fct_collapse(mnd.country, LOW = c("NOR","SWE","FIN","DNK","LUX") , MEDIUM = c("ITA","AUT","DEU","ESP","FRA","NLD"), HIGH = c("GBR","USA","CAN") )

#mnd.country <- fct_collapse(mnd.country, LOW = c("Scandinavia","EasternEurope") , MEDIUM = c("SouthEurope","CentralEurope", "UKIE"), HIGH = c("US") )

# reorder
# NOTE THERE IS NO EASTERN EUROPE
c.mnd.nationality <- fct_collapse(c.mnd.atlas, LOW = c("Scandinavia") , MEDIUM = c("SouthEurope","CentralEurope" ), HIGH = c("US","UKIE") )

c.mnd.nationality <- factor(c.mnd.nationality, levels = (c("LOW", "MEDIUM", "HIGH")))
levels(c.mnd.nationality)

# creates a new variable in the phyloseq called mnd
sample_data(pseq)$mnd = c.mnd.nationality

# checks that it has worked.
get_variable(pseq, "mnd")
table(get_variable(pseq, "mnd"))

###############################################


# THIS IS FOR THE ATLAS 1006 
# Create mnd variable
# grabs the nationality from phyloseq
a.mnd.nationality <- get_variable(atlas.pseq, "nationality" )


# this collapses the two varialbes into a new variable AB in this case. The command for multiple changes is fct_collapse(x, AB = c("A","B"), DE = c("D","E"))
a.mnd.nationality <- fct_collapse(a.mnd.nationality, LOW = c("EasternEurope") , MEDIUM = c("SouthEurope","CentralEurope" ), HIGH = c("US","UKIE","Scandinavia") )

# reorder
a.mnd.nationality <- factor(a.mnd.nationality, levels = (c("LOW", "MEDIUM", "HIGH")))
levels(a.mnd.nationality)

# creates a new variable in the phyloseq called mnd
sample_data(atlas.pseq)$mnd = a.mnd.nationality
#sample_data(atlas.pseq)$country = sample_data(atlas.pseq)$nationality
# checks that it has worked.
table(get_variable(atlas.pseq, "mnd"))


```


##Functions 
```{r function}

# this defines the colours for the graphs. Uses the wes palettes. Look at https://github.com/karthik/wesanderson for the different themes
glob_theme <- "FantasticFox1"




mb.gather <- function (x) {
    # Creates an empty data frame
    mb.group <- data.frame("Location" = c("Central Europe","Eastern Europe","Scandinavia", "Southern Europe","UKIE","USA"), "Mean" = c(1:6)) 
     # groups the countries to match the microbiome classification 
  c.europe <<- filter(x,Location=="Belgium"|Location=="Denmark"|Location=="Germany"|Location=="The Netherlands")
    e.europe <- filter(x,Location=="Poland")
    scan <- filter(x,Location=="Finland"|Location=="Norway"|Location=="Sweden")
    s.europe <- filter(x,Location=="France"|Location=="Italy"|Location=="Serbia"|Location=="Spain")
    ukie <- filter(x,Location=="United Kingdom"|Location=="Ireland")
    usa <- filter(x,Location=="United States")
    # gmeans the data and put it into the data frame
    
    mb.group[1,2] <- round(mean(c.europe$Value),digits = 2)
    mb.group[2,2] <- round(mean(e.europe$Value),digits = 2)
    mb.group[3,2] <- round(mean(scan$Value),digits = 2)
    mb.group[4,2] <- round(mean(s.europe$Value),digits = 2)
    mb.group[5,2] <- round(mean(ukie$Value),digits = 2)
    mb.group[6,2] <- round(mean(usa$Value),digits = 2)

    mb.group[1,3] <- paste(round(mean(c.europe$Upper.bound),digits = 2), round(mean(c.europe$Lower.bound),digits = 2), sep = "-",collapse = "-")
    mb.group[2,3] <- paste(round(mean(e.europe$Upper.bound),digits = 2), round(mean(e.europe$Lower.bound),digits = 2), sep = "-",collapse = "-")
    mb.group[3,3] <- paste(round(mean(scan$Upper.bound),digits = 2), round(mean(scan$Lower.bound),digits = 2), sep = "-",collapse = "-")
    mb.group[4,3] <- paste(round(mean(s.europe$Upper.bound),digits = 2), round(mean(s.europe$Lower.bound),digits = 2), sep = "-",collapse = "-")
    mb.group[5,3] <- paste(round(mean(ukie$Upper.bound),digits = 2), round(mean(ukie$Lower.bound),digits = 2), sep = "-",collapse = "-")
    mb.group[6,3] <- paste(round(mean(usa$Upper.bound),digits = 2), round(mean(usa$Lower.bound),digits = 2), sep = "-",collapse = "-")
    # mb.group[2,3] <- paste(e.europe$Upper.bound,"-", e.europe$Lower.bound)
    # mb.group[3,3] <- paste(scan$Upper.bound,"-", scan$Lower.bound)
    # mb.group[4,3] <- paste(s.europe$Upper.bound,"-",s.europe$Lower.bound)
    # mb.group[5,3] <- paste(ukie$Upper.bound,"-", ukie$Lower.bound )
    # mb.group[6,3] <- paste(usa$Upper.bound,"-", usa$Lower.bound)
    return(mb.group)
    }

mb.cgather <- function (x) {
    # Creates an empty data frame
    mb.group <- data.frame("Location" = c("Norway","Sweden","Finland","Denmark","Luxembourg", "Italy","Austria","Germany","Spain","France","Netherlands","United Kingdom","United States","Canada"), "Mean" = c(1:14)) 
     # groups the countries to match the microbiome classification 
  mb.group <- filter(x,Location=="Norway"|Location=="Sweden"|Location=="Finland"|Location=="Denmark"|Location=="Luxembourg"|Location=="Italy"|Location=="Austria"|Location=="Germany"|Location=="Spain"|Location=="France"|Location=="Netherlands"|Location=="United Kingdom"|Location=="United States"|Location=="Canada")
  
    return(mb.group)
}

my_cur_table <- function(pseq.subset){
    n <- length(sample_data(pseq.subset)$mnd)
    data <- data.frame(id=1:n)
    data$mnd <- sample_data(pseq.subset)$mnd
    data$bmi <- sample_data(pseq.subset)$BMI
    data$bmi_group <- sample_data(pseq.subset)$bmi_group
    data$age <- sample_data(pseq.subset)$age
    data$antibiotics_current_use <- sample_data(pseq.subset)$antibiotics_current_use
    data$smoker <- sample_data(pseq.subset)$smoker
    data$pregnant <- sample_data(pseq.subset)$pregnant
    data$alcohol <- sample_data(pseq.subset)$alcohol
    data$age_category <- sample_data(pseq.subset)$age_category
    table1(~ age  +  bmi + bmi_group + smoker + alcohol + age_category | mnd, data=data)
}

my_atlas_table <- function(atlas.pseq){
    n <- length(sample_data(atlas.pseq)$mnd)
    data <- data.frame(id=1:n)
    data$mnd <- sample_data(atlas.pseq)$mnd
    data$bmi <- sample_data(atlas.pseq)$bmi_group
    data$age <- sample_data(atlas.pseq)$age
    table1(~ age  +  bmi | mnd, data=data)
}

# THis is required
mylab <- function(v1) {
  deparse(substitute(v1))
}
# aplha_d runs the diversity for each country or  groupoing and then compares the three groups
# example is alpha_d(pseq.subset,"TEST_3") this will generate a png file called TEST_3.png in the 
# currect working directory. This can be checked with getwd()
alpha_d <- function(pseq.subset,title_text) {
  # gets the meta data 
  pseq.meta <- meta(pseq.subset)
  # generates the diversites
  tab <- microbiome::alpha(pseq.subset, index = "shannon")
  
  # creates Shannon
  pseq.meta$Shannon <- tab$diversity_shannon
  mnd <- levels(pseq.meta$mnd) # get the variables
  # Make a pairwise list that we want to compare.
  mnd.pairs <- combn(seq_along(mnd), 2, simplify = FALSE, FUN = function(i)mnd[i])
  mnd.pairs <- list( c("LOW","MEDIUM"), c("HIGH","LOW"), c("MEDIUM","HIGH"))
  
  # Compare differences in Shannon index between mnd group of the study subjects
  pseq.meta$mnd <- factor(pseq.meta$mnd, levels = c("LOW","MEDIUM","HIGH"),  exclude = "NA")
  # on to the graphs!!!
  print("The basic alpha diversity analysis has been completed")
  # This country versus nationality
  p1<-   ggviolin(pseq.meta, x = "nationality", y = "Shannon",
                  add = "boxplot", fill = "nationality", palette = "paired") + theme(axis.text.x = element_text(angle = 90)) + theme(legend.position="none") + xlab("")
  print("First graph has completed")
  # the number in the palettes must equal the number of groups in bmi.pairs
  
  p2 <<- ggviolin(data=subset(pseq.meta, !is.na(mnd)), x = "mnd", y = "Shannon", add = "boxplot", fill = "mnd", na.rm = TRUE, palette = wes_palette(glob_theme)) 
  
  p2 <<- p2 + stat_compare_means(comparisons = mnd.pairs) + xlab("MND Prevalence") + theme(legend.position="none")
  
  print("Second graph has completed")
  #this just generates the titles for the output
  title <- ggdraw() + 
    draw_label(
      title_text,
      fontface = 'bold',
      x = 0,
      hjust = 0
    ) +
    theme(
      # add margin on the left of the drawing canvas,
      # so title is aligned with left edge of first plot
      plot.margin = margin(0, 0, 0, 7)
    )
  
  p3 <<-plot_grid(title,p1,p2, align = "v" , ncol = 1, labels = c('','A', 'B'),rel_heights = c(0.1, 1,1.2))
  
  print("Third graph has completed")
  save_plot(paste(title_text,".png", sep = ""),p3,base_height = 11.69 )
  #ggsave("/Users/Nik/Alpha_diversity_curate.png")
}


# THis includes the analysis as well so can take some time to run. the graphs are generated with the label of beta diversity. It requires an input of a title. 
beta_d <- function(pseq.subset,title_text) {
  #NMDS ordination on Bray-Curtis distance
  p <<-  boxplot(list(Low = p.low.d, Medium = p.med.d, High = p.high.d))
 # p1 = plot_ordination(pseq.subset, pseq.nmds, type = "taxa", color = "mnd", title = "Taxa by Phylum") # I dont think that we need this any more. 
  #p1 + facet_wrap(~Phylum, 3)
  pseq.nmds <- ordinate(pseq.subset, "NMDS", "bray")
  p2 <- plot_ordination(pseq.subset, pseq.nmds, type = "sample", color = "mnd")

  p3 <- p2 + scale_colour_manual(values = wes_palette(glob_theme)) + labs(title = "NMDS",color = "MND Prevalence") + theme(legend.position = "none")
  #plot(p1 + scale_colour_manual(values = c("yellow", "orange", "red", "grey")) + labs(color = "MND Prevalence"))

  # This add the facets but I do not think we need that any more. 
  #p3 <- p2 + scale_colour_manual(values = c("yellow", "orange", "red", "grey")) + facet_wrap(~mnd, nrow = 2) + labs(color = "MND Prevalence")
  save_plot(paste("Beta_diversity_",title_text,"_NMDS.png", sep = "" ), p3)
    # MDS
   pseq.mds <<- ordinate(pseq.subset, "MDS", "bray")
   p4 <- plot_ordination(pseq.subset, pseq.mds, type = "sample", color = "mnd")
 #plot(p1 + scale_colour_manual(values = c( "yellow", "orange", "red", "grey")) + labs(color = "MND Prevalence MDS"))

    p5 <- p4 + scale_colour_manual(values = wes_palette(glob_theme)) + labs(title = "MDS",color = "MND Prevalence") + theme(legend.position = "none")
    save_plot(paste("Beta_diversity_",title_text,"_MDS.png", sep = "" ), p5)

   #PCoA for compositional data with Bray-Curtis distances
    #   p4 <- plot_landscape(pseq.rel, method = "NMDS", distance = "bray", colour = "mnd", size = 3) 
    #   p4 <- p4 + scale_color_brewer(palette = "Dark2")+ scale_fill_gradient(low = "#e0ecf4", high = "#6e016b") 
    #p6 <- plot_landscape(pseq.subset, method = "MDS", distance = "bray", col = "mnd", size = 3) 
  #   # PCoA for compositional data with Bray-Curtis distances
   #p6 <- plot_landscape(microbiome::transform(pseq.core, "compositional"),
    #                    method = "PCoA", distance = "bray") +
    #     labs(title = paste("PCoA / Compositional / Bray-Curtis"))
  # 
    tsne_res <- tsne_phyloseq(pseq.subset, distance='bray', perplexity = 5, verbose=1, rng_seed = 3901)
    p7 <- plot_tsne_phyloseq(pseq.subset, tsne_res,color = 'mnd')
    p8 <- p7 + scale_colour_manual(values = wes_palette(glob_theme)) + labs(title = "TSNE",color = "MND Prevalence") + theme(legend.position = "bottom")
    save_plot(paste("Beta_diversity_",title_text,"_MDS.png", sep = "" ), p8)
       #this just generates the titles for the output
    title <- ggdraw() + 
    draw_label(
    title_text,
    fontface = 'bold',
    x = 0,
    hjust = 0
    ) +
    theme(
    # add margin on the left of the drawing canvas,
    # so title is aligned with left edge of first plot
    plot.margin = margin(0, 0, 0, 7)
    )

    p10 <<-plot_grid(title,p3, p5,p8, ncol = 1, labels = c('','A', 'B','C'),rel_heights = c(0.3, 1.5,1.5,2))
    save_plot(paste("Beta_diversity_ALL_",title_text,".png", sep = "" ), p10)
}



pre_perma_d <- function(pseq.subset) { 
  p.low <<- subset_samples(pseq.subset, mnd == "LOW")
  p.low.d <<- divergence(p.low)
  p.med <<- subset_samples(pseq.subset, mnd == "MEDIUM")
  p.med.d <<- divergence(p.med)
  p.high <<- subset_samples(pseq.subset, mnd == "HIGH")
  p.high.d <<- divergence(p.high)
  p <<- boxplot(list(Low = p.low.d, Medium = p.med.d, High = p.high.d))
  
}


# THis calcualtes the group PERMANOVA perma_d(p.low,p.high,"TESTHigh_vesus_Low")
perma_d <- function(x,y,title_text) { # x & y must be either p.low,p.med or p.high
  # Convert to compositional data perma_d(p.low,p.high,"final_highVlow")
  pseqLvH <- merge_phyloseq(x,y) # pseqLvH <- merge_phyloseq(p.low,p.high)
  pseq.rel <- microbiome::transform(pseqLvH, "compositional")
  otu <- abundances(pseq.rel)
  meta <- meta(pseq.rel)
  permanova <- adonis(t(otu) ~ mnd,
               data = meta, permutations=999, method = "bray")
 # P-value
  print(as.data.frame(permanova$aov.tab)["mnd", "Pr(>F)"])  
  dist <- vegdist(t(otu))
  anov <- anova(betadisper(dist, meta$mnd))
  summary(anov)
  tidy(anov)
  coef <- coefficients(permanova)["mnd1",]
  top.coef <- coef[rev(order(abs(coef)))[1:20]]
  par(mar = c(12, 30, 2, 1))
  #KEEP THIS  TO ENSURE THERE IS NOT A MISTAKE IN THE REORDERING BELOW
  #temp <- TeX("Top Genus")
  #png(paste("PERMOVA_",title_text,".png", sep = "" ), width = 450, height = 210,  units     = "mm",
  #res       = 1500)
  #barplot(sort(top.coef), col = rainbow(20), horiz = T, las = 1, main = temp)
  #dev.off() 
  temp <- as.data.frame(top.coef)
  temp$label <- rownames(temp) 
  temp$label <- sub("s_*", "", temp[,2])
  #temp$colour <- ifelse(temp$top.coef < 0, "firebrick1","steelblue")
  p <<- ggplot(temp,aes(reorder(label, top.coef),top.coef,fill = top.coef > 0)) + geom_col(show.legend = FALSE,) + coord_flip() +  xlab('Taxa') + ylab('p value')  + theme_minimal(base_size = 20)  + annotate("text", x = 2.5, y = 0.02, label = "Harmful Taxa") + annotate("text", x = 19, y = -0.01, label = "Beneficial Taxa")  
  save_plot(paste("PERMANOVA_",title_text,".png", sep = "" ), p)
  #ggsave("test_permanova.png", device = "png",width = 297, height = 210, units = "mm")
 } 

# testpseq <- atlas1006 %>%
#    subset_samples(DNA_extraction_method == "r") %>%
#    aggregate_taxa(level = "Phylum") %>%
#    transform(transform = "compositional")
# p <- plot_composition(testpseq, sample.sort = "Firmicutes",
#          otu.sort = "abundance", verbose = TRUE) +
#      scale_fill_manual(values = default_colors("Phylum")[taxa(testpseq)]) 
# 
# comp_d <- function(pseq,subset,title_text) {
#   png(paste(title_text,"Phylum_abundance.png",sep="_"))
#   p <- plot_bar(pseq.subset,"Phylum") + theme(axis.text.x = element_text(face="italic"))
#   print(p)
# 
#   prev.otu <- plot_taxa_prevalence(pseq.subset, "Phylum" , detection = 5)
#   prev.otu + theme(legend.position="none")
#   ggsave(paste(title_text,"phylum_prevalence.png",sep="_"))
#   tab <- plot_taxa_prevalence(pseq.subset, "Phylum", detection = 5)
#   tab + theme(legend.position="none")
#   ggsave(paste(title_text,"Phylum_prevalence_detection.png",sep="_"))
#   
#   pseq.ph <- aggregate_taxa(pseq.subset, "Phylum")
#   pseq.ph.rel <- microbiome::transform(pseq.ph, "compositional")
#   plot.comp.rel <- plot_composition(pseq.ph.rel, x.label = "sample") +
#     theme(legend.position = "bottom") + theme_bw() +
#     theme(axis.text.x=element_blank()) +
#     c + scale_fill_brewer(palette = "Paired") + guides(fill=guide_legend("Phylum"))
#   print(plot.comp.rel)
#   ggsave("phylum_relative_abundance11.png")
#   plot.comp.rel + theme(legend.position="none")
#   ggsave("phylum_relative_abundance12.png") 
#   
#   pseq.ph <- aggregate_taxa(pseq.subset, "Phylum")
#   pseq.ph.rel <- microbiome::transform(pseq.ph, "compositional")
#   plot.comp.rel <- plot_composition(pseq.ph.rel, average_by = "mnd", x.label = "MND   Incidence", plot.type = "barplot") + guides(fill=guide_legend(title="Phylum"))
#   plot.comp.rel + scale_fill_viridis(discrete = TRUE, option = "plasma") 
#   ggsave("phylum_abundance_bar.png")
#   # Top 5
#   top5P.names = sort(tapply(taxa_sums(pseq), tax_table(pseq)[, "Genus"], sum), TRUE)[1:5]
# 
#   top5P = subset_taxa(pseq, Genus %in% names(top5P.names))
# u  pseq.ph.rel <- microbiome::transform(top5P, "compositional")
#   plot.comp.rel <- plot_composition(pseq.ph.rel, average_by = "mnd", x.label = "MND Incidence", plot.type = "barplot") + guides(fill=guide_legend(title="Phylum"))
# 
#   plot.comp.rel + scale_fill_viridis(discrete = TRUE, option = "plasma") 
#   
#   # Top 10
#   top10P.names = sort(tapply(taxa_sums(pseq), tax_table(pseq)[, "Genus"], sum), TRUE)[1:10]
# 
#   top10P = subset_taxa(pseq, Genus %in% names(top10P.names))
#   pseq.ph.rel <- microbiome::transform(top10P, "compositional")
# 
#   plot.comp.rel <- plot_composition(pseq.ph.rel, average_by = "mnd", x.label = "MND Incidence", plot.type = "barplot") + guides(fill=guide_legend(title="Phylum"))
# 
#   plot.comp.rel + scale_fill_viridis(discrete = TRUE, option = "plasma") 
#   # new
#   p.new <- aggregate_taxa(pseq, "Phylum")
#   p.new.rel <- microbiome::transform(p.new, "compositional")
# 
#   plot.p.new.rel <- plot_composition(p.new.rel, average_by = "mnd", x.label = " ", plot.type = "barplot") + guides(fill=guide_legend(title="Phylum")) + theme(legend.text = element_text( face="italic"))
# 
#   plot.p.new.rel + scale_fill_brewer(palette = "Paired") 
#   ggsave("pnew_abundance_bar.png")
#   
#   # Top Phylum Heatmap
#   pseq.ph <- aggregate_taxa(pseq, "Phylum")
# 
#   pseq.ph.rel <- microbiome::transform(pseq.ph, "compositional")
# 
#   plot.comp.rel <- plot_composition(pseq.ph.rel, x.label = "sample", plot.type = "heatmap")
# 
#   plot(plot.comp.rel+ coord_flip()) + theme(axis.text.x=element_blank())
#   ggsave("phylum_histogram.png")
#   
#   #p.new heatmap
#   pseq.n.ph <- aggregate_taxa(pseq, "Phylum")
#   pseq.n.ph.rel <- microbiome::transform(pseq.n.ph, "compositional")
# 
#   plot.n.comp.rel <- plot_composition(pseq.n.ph.rel, x.label = "sample", plot.type = "heatmap")
# 
#   p <- plot.n.comp.rel + coord_flip()
#   p + theme(axis.text.x=element_blank()) + labs(y = "Sample")
#   ggsave("pnew_histogram.png")
#   
#   # Newwork
#   png("network_bray_phylum.png")
#   jg = make_network(pseq, "taxa", "bray", 0.3)
#   plot_network(jg, pseq, "taxa", color = "Phylum", line_weight = 0.4, label = NULL)
#   dev.off()
#   #Community-composition-plots
#   # Create a data table for ggplot
# # agglomerate at phylum level
# pseq_phy <- tax_glom(pseq.subset, taxrank = "Phylum")                  
# 
# # Transform to rel. abundance (or use ps0.ra)
# pseq_phyl <- transform_sample_counts(pseq_phy, function(x) {x/sum(x)}) 
# 
# # Melt to long format for easy ggploting
# pseq_phylum <- psmelt(pseq_phyl) 
#  
#      # filter(Abundance > 0.01)                             # Filter out low abundance taxa
# # Plot - Phylum
#     p.ra.phylum <- ggplot(pseq_phylum, aes(x = sample, y = Abundance, fill = Phylum)) + 
#     geom_bar(stat = "identity", width = 1) +
#     facet_wrap(~mnd, scales = "free_x", nrow = 4, ncol = 7) +
#     theme(axis.text.x = element_blank()) +
#     theme(axis.title.x = element_blank()) +
#     labs(title = "Abundant Phylum (> 1%)")
# 
#   p <- p.ra.phylum + scale_fill_viridis(discrete = TRUE, option = "A")
#   plot(p)
#   ggsave("phylum_abun_mnd.png")
# 
#  # these are other functions required along the way. 
# 
#     # transform counts to n
#   physeq.scale <- transform_sample_counts(physeq, 
#     function(x) {(n * x/sum(x))}
#   )
#   
#   # Pick the rounding functions
#   if (round == "floor"){
#     otu_table(physeq.scale) <- floor(otu_table(physeq.scale))
#   } else if (round == "round"){
#     otu_table(physeq.scale) <- myround(otu_table(physeq.scale))
#   }
#   
#   # Prune taxa and return new phyloseq object
#   physeq.scale <- prune_taxa(taxa_sums(physeq.scale) > 0, physeq.scale)
#   return(physeq.scale)
# }
# 

##### ADONIS ###########

# Function to run adonis test on a phyloseq object and a variable from metadata
# Make sure OTU data is standardized/normalized before 
phyloseq_to_adonis <- function(physeq, distmat = NULL, dist = "bray", formula) {
  
  if(!is.null(distmat)) {
    phydist <- distmat
  } else {
    phydist <- phyloseq::distance(physeq, dist)
  }
  
  metadata <- as(sample_data(physeq), "data.frame")
  
  # Adonis test
  f <- reformulate(formula, response = "phydist")
  adonis.test <- adonis(f, data = metadata)
  print(adonis.test)

  # Run homogeneity of dispersion test if there is only 1 variable
  if (grepl("\\+", formula)) {
    l <- list(
      dist = phydist, 
      formula = f, 
      adonis = adonis.test
    )
  } else {
    group <- metadata[ ,formula]
    beta <- betadisper(phydist, group)
    disper.test = permutest(beta)
    print(disper.test)
    
    l <- list(
      dist = phydist, 
      formula = f, 
      adonis = adonis.test, 
      disper = disper.test
    )
  }
  return (l)
}

########## Bar Plots #################

# This function takes a phyloseq object, agglomerates OTUs to the desired taxonomic rank, 
# prunes out OTUs below a certain relative proportion in a sample (ie 1% ) 
# and melts the phyloseq object into long format which is suitable for ggplot stacked barplots.
taxglom_and_melt <- function(physeq, taxrank, prune){
  
  # Agglomerate all otu's by given taxonomic level
  pglom <- tax_glom(physeq, taxrank = taxrank)
  
  # Create a new phyloseq object which removes taxa from each sample below the prune parameter
  pglom_prune <- transform_sample_counts(pglom,function(x) {x/sum(x)})
  otu_table(pglom_prune)[otu_table(pglom_prune) < prune] <- 0
  pglom_prune <- prune_taxa(taxa_sums(pglom_prune) > 0, pglom_prune)
  
  # Melt into long format and sort by taxonomy
  physeq_long <- psmelt(pglom_prune)
  physeq_long <- physeq_long[order(physeq_long[ ,taxrank]), ]

  # Return long data frame
  return(physeq_long)
}


###### Merge functions ############

# Merge samples by averaging OTU counts instead of summing
merge_samples_mean <- function(physeq, group, round){
  # Calculate the number of samples in each group
  group_sums <- as.matrix(table(sample_data(physeq)[ ,group]))[,1]
  
  # Merge samples by summing
  merged <- merge_samples(physeq, group)
  
  # Divide summed OTU counts by number of samples in each group to get mean
  # Calculation is done while taxa are columns
  x <- as.matrix(otu_table(merged))
  if(taxa_are_rows(merged)){ x<-t(x) }

  # Pick the rounding functions
  if (round == "floor"){
    out <- floor(t(x/group_sums))
  } else if (round == "round"){
    out <- myround(t(x/group_sums))
  }
  
  # Return new phyloseq object with taxa as rows
  out <- otu_table(out, taxa_are_rows = TRUE)
  otu_table(merged) <- out
  return(merged)
}

# Merge samples, just including OTUs that were present in all merged samples
# Call this function before running merge_samples()
merge_OTU_intersect <- function(physeq, group){
  
  # Make sure we're not starting with more taxa than we need 
  physeq <- prune_taxa(taxa_sums(physeq) > 0, physeq)
  
  s <- data.frame(sample_data(physeq))
  l <- levels(s[,group])
  o <- otu_table(physeq)
  
  # Loop through each category
  for (cat in 1:length(l)) {
 
    # Get the index of all samples in that category
    w <- which(s[,group]==l[cat])
   
    # subset to just those columns of OTU table
    cat.sub<-o[,w]
    print(dim(cat.sub))
    
    # Find the indices of 0's in the OTU table
    zeros <- apply(cat.sub, 1, function(r) any(r == 0))
    
    # If an OTU had a 0 in at least one sample, change all samples to 0
    cat.sub[zeros,] <- 0
  }
  
  o[,w] <- cat.sub
  otu_table(physeq) <- o
  
  return(physeq)
  
}




```


```{r check OTUs}
# Check if any OTUs are not present in any samples
any(taxa_sums(pseq.subset) == 0)
# FALSE, therefore, OTUs are found in all samples and we don't need to prune the taxa. If the answer is TRUE, some OTUs are not present in any samples. This is usually the case when data is subset to remove some samples. OTUs unique to those sample are not removed along with the samples. Therefore, it is important to check this everytime the phyloseq object is filtered for samples using `subset_samples` function.

head(meta(pseq))

```

## Filtering
```{r  data filtering}

# https://bioconductor.riken.jp/packages/3.5/data/experiment/vignettes/curatedMetagenomicData/inst/doc/curatedMetagenomicData.html
# Prepare data for vizualisation

# Rmove   everything without a country

keepotu = genefilter_sample(pseq, filterfun_sample(topp(0.05)), A=5)
pseq.f <- subset_taxa(pseq, keepotu)

pseq.c <- core(pseq, detection = 0.1/100, prevalence = 50/100)

#pseq.subset <- subset_samples(pseq.c, !is.na(mnd) & disease == "healthy" & age_category != "child" & age_category != "newborn" & age_category != "schoolage" & antibiotics_current_use != "yes")

pseq.subset <- subset_samples(pseq.c, !is.na(mnd) & disease == "healthy" & age_category != "child" & age_category != "newborn" & age_category != "schoolage" )


setwd("~/globalmicrobiome") 

# These are useful is filtering. 
#sample_variables(pseq.subset)
#get_variable(pseq.subset, sample_variables(pseq.subset)[13])
my_cur_table(pseq.subset)
my_cur_table(pseq.c)

my_atlas_table(atlas.pseq)
alpha_d(pseq,"final_RLWT_1")

alpha_d(atlas.pseq,"final_RLWT_2")

beta_d(pseq.subset,"final_2")

pre_perma_d(pseq.subset) 
perma_d(p.low,p.high,"RLWT_Curatefinal_highVlow")

pre_perma_d(pseq.subset) 
perma_d(p.med,p.high,"RLWT_Curatefinal_highVmed")


pre_perma_d(atlas.pseq) 
perma_d(p.low,p.high,"RLWT_Atlasfinal_highVlow")

pre_perma_d(atlas.pseq) 
perma_d(p.med,p.high,"RLWT_Atlasfinal_highVmed")




#   not working from here !!!!!!

plot_abundance.ak = function(pseq.subset,title = "",
                         Facet = "Genus", Color = "Phylum"){
 # Arbitrary subset, based on Phylum, for plotting
 p1f = subset_taxa(pseq.subset, Genus %in% c("Akkermansia"))
 mphyseq = psmelt(p1f)
 mphyseq <- subset(mphyseq, Abundance > 0)
 p <- ggplot(data = mphyseq, mapping = aes_string(x = "mnd",y = "Abundance")) +
   geom_violin(fill = NA) +
   geom_point(size = 1, alpha = 0.3,
              position = position_jitter(width = 0.3)) + scale_y_log10()+
   theme(legend.position="none") + scale_colour_manual(values = wes_palette(glob_theme)) 
 library(rms)
fita <- lrm(mnd ~ gender + log(Abundance) + bmi_group + age_category, data =mphyseq)
fita <- lrm(mnd ~ gender + log(Abundance) , data =mphyseq)
fita <- lrm(mnd ~ gender + log(Abundance) + bmi_group, data =mphyseq)

}
 

newData<-data.frame(JVeg5=factor(Jdata[,"JVeg5"]),scale(Jdata[,c("Elevation","Lat_Y_pos","Coast_dist","Stream_dist")]))

Global <- polr(JVeg5 ~  Elevation +  Lat_Y_pos + Coast_dist + Stream_dist,
                data=newData, na.action = na.omit, Hess = TRUE)

library(MASS)
model_fit <- polr(mnd~Abundance, data = mphyseq)
summary(model_fit)


data(atlas1006)
plot_regression(diversity ~ mnd, meta(atlas.pseq))

otu <- atlas.pseq$microbes 
lipids <- peerj32$lipids 

# Define data sets to cross-correlate
x <- log10(otu) # OTU Log10 (44 samples x 130 genera)
y <- as.matrix(lipids) # Lipids (44 samples x 389 lipids)

# Cross correlate data sets
correlations <- associate(x, y, method = "spearman", mode = "matrix", p.adj.threshold = 0.05, n.signif = 1)

# Or, alternatively, the same output is also available in a handy table format
correlation.table <- associate(x, y, method = "spearman", mode = "table", p.adj.threshold = 0.05, n.signif = 1)

kable(head(correlation.table))
```


```{r MNDtable summary}





atlas.pseq
meta <- meta(atlas.pseq)



```


## BETA DIVERSITY
Beta-diversity: Measures for differences between samples from different groups to identify if there are differences in the overall community composition and structure.
WE want to: quantify community divergence within a given sample set
Divergence of a given sample can be quantified as the average dissimilarity of each sample from the group mean; the dissimilarity can be quantified by beta diversity.

Divergence of a given sample set can be quantified as the average dissimilarity of each sample from the group mean; the dissimilarity can be quantified by beta diversity, for instance. This was applied in group-level comparisons for instance in Salonen et al. ISME J 2014 (they focused on homogeneity using inverse correlation, whereas here we focus on divergence using correlation but the measure is essentially the same).

 The low group tends to have smaller values, indicating that the samples are more similar to the group mean, and is less heterogeneous (has smaller spread / is more homogeneous):

## PERMANOVA significance test for group-level differences
Evaluate whether the MND grouping has a significant effect on overall gut microbiota composition





NOTHING BELOW THIS WORKS!!!!!!!!






## MICROBIOME COMPOSITION

```{r Phylum #}

# Show available ranks in the dataset
rank_names(pseq)

# Create table, number of features for each phyla
table(tax_table(pseq)[, "Phylum"], exclude = NULL)

```

This shows a few phyla for which only one feature was observed. Those may be worth filtering.


guide_italics
ggtitle("Relative abundance") +  
```{r}

clos.ph <- aggregate_taxa(clostridium, "Phylum")

clos.ph.rel <- microbiome::transform(clos.ph, "compositional")

plot.comp.rel <- plot_composition(clos.ph.rel, x.label = "sample") + 
  theme(legend.position = "bottom", ) + theme_bw() + 
  theme(axis.text.x=element_blank()) + 
  theme(legend.title = element_text(size=18), legend.text = element_text( face="italic")) + scale_fill_brewer(palette = "Paired") + guides(fill=guide_legend(title="Phylum")) 

print(plot.comp.rel)

ggsave("phylum_relative_abundance_clos.png")

plot.comp.rel + scale_fill_brewer(palette = "Paired") + theme(legend.position="none") + guide_italics
ggsave("phylum_relative_abundance_clos2.png")       



```




```{r CLostridium Bar Plot}

clos <- aggregate_taxa(clostridium, "Phylum")

pseq.cl.rel <- microbiome::transform(clos, "compositional")

plot.cl.rel <- plot_composition(pseq.cl.rel, average_by = "mnd", x.label = "sample", plot.type = "barplot") + guides(fill=guide_legend(title="Phylum")) + theme(legend.text = element_text( face="italic")) 

plot.cl.rel + scale_fill_brewer(palette = "Paired")
ggsave("cl_abundance_bar.png")

```

```{r Clostridium genus bar plots}

pseq.cl.rel <- microbiome::transform(clostridium, "compositional")

plot.cl.rel <- plot_composition(pseq.cl.rel, average_by = "mnd", x.label = "sample", plot.type = "barplot")

plot.cl.rel + scale_fill_viridis(discrete = TRUE, option = "A") 

```



```{r clostridium heatmap}

pseq.c.ph <- aggregate_taxa(clostridium, "Phylum")

pseq.c.ph.rel <- microbiome::transform(pseq.c.ph, "compositional")

plot.c.comp.rel <- plot_composition(pseq.c.ph.rel, x.label = "sample", plot.type = "heatmap")

p <- plot.c.comp.rel + coord_flip()
p + theme(axis.text.x=element_blank()) + labs(y = "Sample")
ggsave("clos_histogram.png")

```






## Community composition plotting
Classic bar plots of bacterial phyla present per sample can be useful for communicating "high level" results. These are relatively easy to interpret when major shifts in microbial communities are present, such as in this study where antibiotics are used However, they are not effective at detecting subtle shifts in communities or taxa and do not convey any statistical significance and can be subjectively interpreted. Interpretation of these plots should always be subject to subsequent statistical analysis.

```{r}

clos <- aggregate_taxa(clostridium, "Phylum")

# agglomerate at phylum level
clos_phy <- tax_glom(clos, taxrank = "Phylum")                  

# Transform to rel. abundance (or use ps0.ra)
clos_phyl <- transform_sample_counts(clos_phy, function(x) {x/sum(x)}) 

# Melt to long format for easy ggploting
clos_phylum <- psmelt(clos_phyl)

# Plot - Phylum
p.ra.phylum <- ggplot(clos_phylum, aes(x = sample, y = Abundance, fill = Phylum)) + 
  geom_bar(stat = "identity", width = 1) +
  facet_wrap(~mnd, scales = "free_x", nrow = 4, ncol = 7) +
  theme(axis.text.x = element_blank()) +
  theme(axis.title.x = element_blank()) +
  labs(title = "Abundant Phylum of Clostridium(> 1%)")

p <- p.ra.phylum + scale_fill_viridis(discrete = TRUE, option = "A")
plot(p)
ggsave("phylum_abun_mnd_clos.png")

```



```{r}

get_taxa_unique(pseq, "Phylum") 

```


```{r}

get_taxa_unique(pseq, "Genus") 

```

```{r}

get_taxa_unique(clostridium, "Genus") 

```


```{r}

df <- psmelt(pseq)
kable(head(df))

```

```{r}

head(abundances(pseq)["Faecalibacterium prausnitzii et rel.",])

```



