---
title: "R Notebook"
output: html_notebook
---

This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook. When you execute code within the notebook, the results appear beneath the code. 


BiocManager::install("curatedMetagenomicData")

BiocManager::install("ExperimentHub")
library(ExperimentHub)
library(curatedMetagenomicData)



Combining 2 dat asets

mergedGP = merge_samples(TwinsUK,atlas1006)

ps_merge <- merge_phyloseq(atlas1006, tax_table(TwinsUK))

ps_merge <- merge_phyloseq(TwinsUK, atlas1006)

 get_variable(batmicrobiome,""LATITUDE"" )
 
 get_variable(batmicrobiome,"LONGITUDE" )
 
sample_data(atlas1006)
 
 sample_data(atlas1006)
 
 data=subset(pseq.meta, !is.na(mnd))
 
sample_names(batmicrobiome)

batmicrobiome[LONGITUDE] <- sample(50:1000)

pseq_temp <- baseline(atlas1006)

temp <- complete.cases(sample_data(atlas1006))

sample_data(atlas1006)$human = factor( get_variable(atlas1006) %in% c("Feces", "Mock", "Skin", "Tongue") )

sample_data(atlas1006)$human = sample(50:1000)

table(sample_data(atlas1006)$nationality)



sample_data(atlas1006)$human = rep(3,1151) %%% THIS WORKS !!!!


sample_data(atlas1006)$LATITUDE = rep(17,1151)
sample_data(atlas1006)$LONGITUDE = rep(-77,1151) 
get_variable(atlas1006,"LATITUDE" )
map_phyloseq(atlas1006)

map_phyloseq(batmicrobiome)

rep(3,12)

devtools::install_github("UrbanInstitute/urbnmapr")

atlas1006

sample_variables(atlas1006)

sample_variables(batmicrobiome)[56]  THIS WORKS

batmicrobiome = sample_data(data.frame(
  LONGITUDE = sample(50:1000)))
  
  
    sampledata = sample_data(data.frame(
  Location = sample(LETTERS[1:4], size=nsamples(physeq), replace=TRUE),
  Depth = sample(50:1000, size=nsamples(physeq), replace=TRUE),
  row.names=sample_names(physeq),
  stringsAsFactors=FALSE
  
  sampledata = sample_data(data.frame(
  Location = sample(LETTERS[1:4], size=nsamples(physeq), replace=TRUE),
  Depth = sample(50:1000, size=nsamples(physeq), replace=TRUE),
  row.names=sample_names(physeq),
  stringsAsFactors=FALSE
))




#
data(esophagus)
tree <- phy_tree(esophagus)
otu  <- otu_table(esophagus)
otutree0 <- phyloseq(otu, tree)
 plot_tree(otutree0)
otutree1 <- merge_taxa(otutree0, 1:8, 2)
plot_tree(esophagus, ladderize="left")


clostridium <- subset_taxa(pseq, Phylum =="Clostridium cluster I"| "Clostridium cluster XI"|"Clostridium cluster XVI"| "Clostridium cluster III"| "Clostridium cluster XIII"| "Clostridium cluster XVII"| "Clostridium cluster IV"| "Clostridium cluster XVIa"| "Clostridium cluster XVIII" |"Clostridium cluster IX"| "Clostridium cluster XV"| "Uncultured Clostridiales")

stomach <- subset_samples(physeq3, Sample_type=="Stomach")
data(atlas1006)   
# Rename the example data (which is a phyloseq object)
pseq <- atlas1006
x <- tax_glom(pseq, taxrank="Phylum")
pseq <- tax_glom(pseq, taxrank="Genus")

rank_names(pseq)
library(HITChipDB)
 data.dir <- system.file("HITChip.tab", package = "microbiome")
  dat <- import_hitchip(data.dir)
  
  pseq <- HITChipDB::read_hitchip(data.dir, method = "frpa")$pseq
  
  treefile = read_tree("HITdb.tre", errorIfNULL = T)
  
    library(microbiome)
  data(peerj32)
  otu <- peerj32$microbes
  meta <- peerj32$meta
  
  # drops it into different countries
  pseq.subset <- subset_samples(pseq, nationality == "SouthEurope"& gender=="female" )
  
  clostridium <- subset_samples(pseq, Phylum =="Clostridium")
  
  library(car)
recode(sample_data(pseq)$nationality, "c('SouthEurope', 'CentralEurope')='A+B'")
  
  
  
  df %>%
   mutate(x= recode(sample_data(pseq)$nationality, "c('SouthEurope', 'CentralEurope')='A+B'"))
  
sample_data(pseq)$bob[subset_samples(pseq, nationality == "SouthEurope"& nationality == "Scandinavia" )] <- "TEST"

# grabs the nationality from phyloseq
test <- get_variable(pseq,"nationality" ) 
# check you've installed this library
library(forcats)
# this collapses the two varialbes into a new variable AB in this case. The command for multiple changes is fct_collapse(x, AB = c("A","B"), DE = c("D","E"))
test <- fct_collapse(test, AB = c("SouthEurope","CentralEurope"))
# creates a new variable in the phyloseq called mnd
sample_data(pseq)$mnd = test
# checks that it has worked. 
get_variable(pseq,"mnd" )

data$grip6max[data$grprespd==3 & data$sex==1] <- 29.46 

# Gives an overview of the data set
sample_data(cdi_schubert)
sample_data(ps_merge)
sample_data(pseq)$nationality

https://www.nature.com/articles/s41467-017-01973-8.pdf?origin=ppub

dietswap

ps_merge <- merge_phyloseq(cdi_schubert, atlas1006)

)
  
Try executing this chunk by clicking the *Run* button within the chunk or by placing your cursor inside it and pressing *Cmd+Shift+Enter*. 

```{r}
	setwd("/omega/bashscripts/microbiomedatarepo/datasets/microbiomeDB") 


	## Get data
	library(foreign)
	library(stargazer)
	library(psych)
	library(plyr)
	library(xtable)
	library(car)
	#library(dplyr)
	library(scales)
	library(reshape) 
	library(gmodels)
	library(nnet)
	library(tidyr)
	library(sjmisc)
	library(RColorBrewer)
  library(htmltools)
	library(broom)

library(hrbrthemes)
library(gcookbook)
library(tidyverse)

library(phyloseq)
library(microbiome)
library(knitr)
library(ggpubr)

library(phylogeo)
library("HITChipDB")
library(phangorn)
library(microbiomeutilities)


data(atlas1006)   
# Rename the example data (which is a phyloseq object)
pseq <- atlas1006
#pseq <- tax_glom(pseq, taxrank="Phylum")
#pseq <- tax_glom(pseq, taxrank="Genus")
```

Taxonomic Filtering

In many biological settings, the set of all organisms from all samples are well-represented in the available taxonomic reference database. When (and only when) this is the case, it is reasonable or even advisable to filter taxonomic features for which a high-rank taxonomy could not be assigned. Such ambiguous features in this setting are almost always sequence artifacts that don’t exist in nature. It should be obvious that such a filter is not appropriate for samples from poorly characterized or novel specimens, at least until the possibility of taxonomic novelty can be satisfactorily rejected. Phylum is a useful taxonomic rank to consider using for this purpose, but others may work effectively for your data.

```{r}

table(tax_table(pseq)[, "Phylum"], exclude = NULL)

# Compute prevalence of each feature, store as data.frame
prevdf = apply(X = otu_table(pseq),
               MARGIN = ifelse(taxa_are_rows(pseq), yes = 1, no = 2),
               FUN = function(x){sum(x > 0)})

# Add taxonomy and total read counts to this data.frame
prevdf = data.frame(Prevalence = prevdf,
                    TotalAbundance = taxa_sums(pseq),
                    tax_table(pseq))

# Are there phyla that are comprised of mostly low-prevalence features? Compute the total and average prevalences of the features in each phylum.
tab <- plyr::ddply(prevdf, "Phylum", function(df1){cbind(mean(df1$Prevalence),sum(df1$Prevalence))})

# Plot
stargazer(tab, summary=FALSE, type = "text")

#IF NEED TO REMOVE
# Define phyla to filter
#filterPhyla = c("Fusobacteria", "Deinococcus-Thermus")
# Filter entries with unidentified Phylum.
#ps1 = subset_taxa(ps, !Phylum %in% filterPhyla)
#ps1

```


Prevalence Filtering
The previous filtering steps are considered supervised, because they relied on prior information that is external to this experiment (a taxonomic reference database). This next filtering step is completely unsupervised, relying only on the data in this experiment, and a parameter that we will choose after exploring the data. Thus, this filtering step can be applied even in settings where taxonomic annotation is unavailable or unreliable.

First, explore the relationship of prevalence and total read count for each feature. Sometimes this reveals outliers that should probably be removed, and also provides insight into the ranges of either feature that might be useful. This aspect depends quite a lot on the experimental design and goals of the downstream inference, so keep these in mind. It may even be the case that different types of downstream inference require different choices here. There is no reason to expect ahead of time that a single filtering workflow is appropriate for all analysis.

Each point in Figure 2 is a different taxa. Exploration of the data in this way is often useful for selecting filtering parameters, like the minimum prevalence criteria we will used to filter the data above.

Sometimes a natural separation in the dataset reveals itself, or at least, a conservative choice that is in a stable region for which small changes to the choice would have minor or no effect on the biological interpreation (stability). Here no natural separation is immediately evident, but it looks like we might reasonably define a prevalence threshold in a range of zero to ten percent or so. Take care that this choice does not introduce bias into a downstream analysis of association of differential abundance.

The following uses five percent of all samples as the prevalence threshold.

```{r}
# Subset to the remaining phyla
prevdf1 = subset(prevdf, Phylum %in% get_taxa_unique(pseq))
ggplot(prevdf1, aes(TotalAbundance, Prevalence / nsamples(pseq),color=Phylum)) +
  # Include a guess for parameter
  geom_hline(yintercept = 0.05, alpha = 0.5, linetype = 2) +  geom_point(size = 2, alpha = 0.7) +
  scale_x_log10() +  xlab("Total Abundance") + ylab("Prevalence [Frac. Samples]") +
  facet_wrap(~Phylum) + theme(legend.position="none")
```

The following uses five percent of all samples as the prevalence threshold.

```{r}
# Define prevalence threshold as 5% of total samples
prevalenceThreshold = 0.05 * nsamples(pseq)
keepTaxa = rownames(prevdf1)[(prevdf1$Prevalence >= prevalenceThreshold)]
ps2 = prune_taxa(keepTaxa, pseq)
```

Agglomerate taxa
When there is known to be a lot of species or sub-species functional redundancy in a microbial community, it might be useful to agglomerate the data features corresponding to closely related taxa. Ideally we would know the functional redundancies perfectly ahead of time, in which case we would agglomerate taxa using those defined relationships and the function in phyloseq. That kind of exquisite functional data is usually not available, and different pairs of microbes will have different sets of overlapping functions, complicating the matter of defining appropriate grouping criteria.

While not necessarily the most useful or functionally-accurate criteria for grouping microbial features (sometimes far from accurate), taxonomic agglomeration has the advantage of being much easier to define ahead of time. This is because taxonomies are usually defined with a comparatively simple tree-like graph structure that has a fixed number of internal nodes, called “ranks”. This structure is simple enough for the phyloseq package to represent taxonomies as table of taxonomy labels. Taxonomic agglomeration groups all the “leaves” in the hierarchy that descend from the user-prescribed agglomerating rank, this is sometimes called ‘glomming’.

The following example code shows how one would combine all features that descend from the same genus.

```{r}
# How many genera would be present after filtering?
length(get_taxa_unique(ps2, taxonomic.rank = "Genus"))
## [1] 49
ps3 = tax_glom(ps2, "Genus", NArm = TRUE)
ps4 = tax_glom(ps2, "Phylum", NArm = TRUE)

#Here phyloseq’s plot_tree() function compare the original unfiltered data, the tree after taxonoic agglomeration, and the tree after phylogenetic agglomeration. These are stored as separate plot objects, then rendered together in one combined graphic using  gridExtra::grid.arrange.

multiPlotTitleTextSize = 15
p2tree = plot_tree(ps2, method = "treeonly",
                   ladderize = "left",
                   title = "Before Agglomeration") +
  theme(plot.title = element_text(size = multiPlotTitleTextSize))
p3tree = plot_tree(ps3, method = "treeonly",
                   ladderize = "left", title = "By Genus") +
  theme(plot.title = element_text(size = multiPlotTitleTextSize))
p4tree = plot_tree(ps4, method = "treeonly",
                   ladderize = "left", title = "By Phylum") +
  theme(plot.title = element_text(size = multiPlotTitleTextSize))

# group plots together
grid.arrange(nrow = 1, p2tree, p3tree, p4tree)

```

```{r}
library(dada2)
library(DECIPHER)
ref_fasta <- "data/rdp_train_set_14.fa.gz"
taxtab <- assignTaxonomy(seqtab, refFasta = ref_fasta)
colnames(taxtab) <- c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus")

taxonomy <- tax_table(pseq)

```


Add a new chunk by clicking the *Insert Chunk* button on the toolbar or by pressing *Cmd+Option+I*.

```{r}



tab <- global(pseq, index = "all")
stargazer(tab, type = "text")

```


```{r}
####### Alpha diversity #######
ps1 <- prune_taxa(taxa_sums(pseq) > 0, pseq)
tab <- diversities(ps1, index = "all")
stargazer(tab, type = "text",summary = NULL)
kable(head(tab))


alpha2 <- . %>%
  prune_taxa(taxa_sums(.) > 0, .) %>%
  diversities(., index = "all")  %>%
  kable(head(.))

  
```

```{r}
#  Prepare data for vizualisation
ps1.meta <- meta(ps1)
kable(head(ps1.meta))

stargazer(ps1.meta, type = "text",summary = NULL)
# Add the diversity table to metadata
ps1.meta$Shannon <- tab$shannon 
ps1.meta$InverseSimpson <- tab$inverse_simpson

# Compare differences in Shannon index between bmi group of the study subjects

# create a list of pairwise comparisons
bmi <- levels(ps1.meta$bmi_group) # get the variables

# make a pairwise list that we want to compare.
bmi.pairs <- combn(seq_along(bmi), 2, simplify = FALSE, FUN = function(i)bmi[i])

print(bmi.pairs)
```


```{r}
# CREATING VIOLIN PLOT
# the number in the palettes must equal the number of groups in bmi.pairs
p1 <- ggviolin(ps1.meta, x = "bmi_group", y = "Shannon",
 add = "boxplot", fill = "bmi_group", palette = c("#a6cee3", "#b2df8a", "#fdbf6f","#a6cee3", "#b2df8a", "#fdbf6f")) # 
print(p1)
```



```{r}

# Pairwise comparision using non-parametric test (Wilcoxon test).
p1 <- p1 + stat_compare_means(comparisons = bmi.pairs) 
print(p1)
```


```{r}
####### RICHNESS #######

tab <- richness(pseq)
kable(head(tab))
stargazer(tab, type = "text",summary = NULL)
plot(tab)

```



```{r}
# CORE

# Calculate compositional version of the data
# (relative abundances)
pseq.rel <- microbiome::transform(pseq, "compositional")


#Relative population frequencies; at 1% compositional abundance threshold:

head(prevalence(pseq.rel, detection = 1, sort = TRUE))

# Absolute population frequencies (sample count):

head(prevalence(pseq.rel, detection = 1, sort = TRUE, count = TRUE))


# This returns the taxa that exceed the given prevalence and detection thresholds
pseq.core <- core(pseq.rel, detection = 0, prevalence = .5)

# Retrieving the associated taxa names from the phyloseq object:
core.taxa <- taxa(pseq.core)


# Total core abundance in each sample (sum of abundances of the core members):
core.abundance <- sample_sums(core(pseq.rel, detection = .01, prevalence = .95))

# With compositional (relative) abundances
det <- c(0, 0.1, 0.5, 2, 5, 20)/100
prevalences <- seq(.05, 1, .05)
plot_core(pseq.rel, prevalences = prevalences, detections = det, plot.type = "lineplot") + xlab("Relative Abundance (%)")

# COREHEAT MAPS

# Core with compositionals:
prevalences <- seq(.05, 1, .05)
detections <- 10^seq(log10(1e-3), log10(.2), length = 10)

# Also define gray color palette
gray <- gray(seq(0,1,length=5))
p <- plot_core(pseq.rel, plot.type = "heatmap", colours = gray,
    prevalences = prevalences, detections = detections) +
    xlab("Detection Threshold (Relative Abundance (%))")
print(p)    


# Same with the viridis color palette
# color-blind friendly and uniform
# options: viridis, magma, plasma, inferno
# https://cran.r-project.org/web/packages/viridis/vignettes/intro-to-viridis.html
# Also discrete=TRUE versions available
library(viridis)

## Loading required package: viridisLite

print(p + scale_fill_viridis())

## Scale for 'fill' is already present. Adding another scale for 'fill',
## which will replace the existing scale.

# Core with absolute counts and horizontal view:
# and minimum population prevalence (given as percentage)
detections <- 10^seq(log10(1), log10(max(abundances(pseq))/10), length = 10)

library(RColorBrewer)
p <- plot_core(pseq, plot.type = "heatmap", 
             prevalences = prevalences,
             detections = detections,
         colours = rev(brewer.pal(5, "Spectral")),
         min.prevalence = .2, horizontal = TRUE)
print(p)


```


```{r}
###### DOMINANCE #####

# The dominance index refers to the abundance of the most abundant species. Various dominance indices are available (see the function help for a list of options). 


# Absolute abundances for the single most abundant taxa in each sample
tab <- dominance(pseq, index = "all")
kable(head(tab))
stargazer(tab, type = "text",summary = NULL)
plot(tab)

p <- ggplot(tab) + 
```


```{r}
######### RARENESS and low abundance ######

tab <- rarity(pseq, index = "all")
kable(head(tab))
stargazer(tab, type = "text",summary = NULL)
plot(tab)
```

```{r}
#### COVERAGE 

# The coverage index gives the number of groups needed to have a given proportion of the ecosystem occupied (by default 0.5 ie 50%).

tab <- coverage(pseq, threshold = 0.5)
kable(head(tab))
stargazer(tab, type = "text",summary = NULL)
plot(tab)
```


```{r}
#### ABUNDANCE 

# The core_abundance function refers to the relative proportion of the core species. Non-core abundance provides the complement (1-x; see noncore_abundance).

tab <- core_abundance(pseq, detection = .1/100, prevalence = 50/100)
kable(head(tab))
stargazer(tab, type = "text",summary = NULL)
plot(tab)

```

```{r}
##### GINI INDEX

#Gini index is a common measure for inequality in economical income. The inverse gini index (1/x) can also be used as a community diversity measure.

tab <- inequality(pseq)
kable(head(tab))
plot(tab)
```

```{r}
###### EVENNESS

# Various evenness measures are also available.

tab <- evenness(pseq, "all")
kable(head(tab))
plot(tab)
```

MICROBIOME COMPOSITION 


```{r}
# Just use prevalent taxa to speed up examples
# (not absolute counts used in this example)

# Pick sample subset
library(phyloseq)

library(hrbrthemes)
library(gcookbook)
library(tidyverse)
library(dplyr)
library(microbiome)


######## THIS ALTERED THE CORE DATA ###############
#pseq <- core(atlas1006, detection = 10^3, prevalence = 95/100)
#pseq2 <- subset_samples(pseq, group == "DI" & nationality == "AFR" & timepoint.within.group == 1)

# Try another theme
# from https://github.com/hrbrmstr/hrbrthemes

# Limit the analysis on core taxa and specific sample group
p <- plot_composition(pseq,
              taxonomic.level = "OTU",
                      sample.sort = "nationality",
                      x.label = "nationality") +
     guides(fill = guide_legend(ncol = 1)) +
     scale_y_percent() +
     labs(x = "Samples", y = "Relative abundance (%)",
                                   title = "Relative abundance data",
                                   subtitle = "Subtitle",
                                   caption = "Caption text.") + 
     theme_ipsum(grid="Y")
print(p)  

# Averaged by group
p <- plot_composition(pseq,
                      average_by = "bmi_group", transform = "compositional")
print(p)


```


```{r}
tmp <- plot_composition(pseq, plot.type = "heatmap", transform = "compositional", 
            sample.sort = "neatmap", otu.sort = "neatmap", mar = c(6, 13, 1, 1))

print(tmp)
```





PLOTTING TAXA Prevlence 

```{r}
data(atlas1006)

# Use sample and taxa subset to speed up example
p0 <- subset_samples(atlas1006, DNA_extraction_method == "r")

# Define detection and prevalence thresholds to filter out rare taxa
p0 <- core(p0, detection = 10, prevalence = 0)

# For the available taxonomic levels
plot_taxa_prevalence(p0, "Phylum", detection = 10)
```


PLOTTING RELAVANCE ABUNDANCE

```{r}

library(phyloseq)
ps2 <- data(atlas1006)
# for example purpose we will remove samples with less than 10000
ps2 = prune_samples(sample_sums(ps1)>=2000, ps1)

# To speed up the example we will use only those OTUs that are detected 100 times and present in 50% of the samples.

pseq2 <- microbiome::core(ps2, detection = 100, prevalence = .5)

# Improve the plotting MODIFIED 
tax_table(pseq2)[tax_table(pseq2)[,"Phylum"]== "Clostridium", "Phylum" ] <- "Clostridium Phylum"

# We will also remove the "f__" patterns for cleaner labels
tax_table(pseq2)[,colnames(tax_table(pseq2))] <- gsub(tax_table(pseq2)[,colnames(tax_table(pseq2))],pattern="[a-z]__",replacement="")

# merge at family level.

pseq.fam <- aggregate_taxa(pseq2, "Phylum")

p.fam <- plot_composition(pseq.fam, sample.sort = NULL, otu.sort = NULL,
  x.label = "ibd_subtype", plot.type = "barplot", verbose = FALSE)

print(p.fam)

```


PLOTTING RELTIVE ABUNDANCE

```{r}

### REQUIREDS THE DATA FROM ABOVE 
pseq.famrel <- transform(pseq.fam, "compositional")

p.famrel <- plot_composition(pseq.famrel, sample.sort = NULL, otu.sort = NULL,
  x.label = "ibd_subtype", plot.type = "barplot", verbose = FALSE)

print(p.famrel)

# further improvements can be done as follows  

p.famrel <- plot_composition(pseq.famrel, 
                             sample.sort = NULL, 
                             otu.sort = NULL, 
                             x.label = "ibd_subtype", 
                             plot.type = "barplot", 
                             verbose = FALSE) + 
  guides(fill = guide_legend(ncol = 1)) + 
  scale_y_percent() + 
  labs(x = "Samples", 
       y = "Relative abundance (%)",
       title = "Relative abundance data", 
       subtitle = "Subtitle",
       caption = "Caption text.") + 
     theme_ipsum(grid="Y")

print(p.famrel)
```

HEAT MAP

```{r}
pseq.famlog <- transform(pseq.fam, "log10")

p.famrel.heatmap <- plot_composition(pseq.famlog, 
                             sample.sort = NULL, 
                             otu.sort = NULL, 
                             x.label = "ibd_subtype", 
                             plot.type = "heatmap", 
                             verbose = FALSE)

print(p.famrel.heatmap)
```

Plot taxa prevalence

```{r}

library(microbiome)
data(peerj32)

# Rename the data
pseq <- peerj32$phyloseq

# Calculate compositional version of the data
# (relative abundances)
pseq.rel <- microbiome::transform(pseq, "compositional")


core.taxa.standard <- core_members(pseq.rel, detection = 0, prevalence = 50/100)


```


When you save the notebook, an HTML file containing the code and output will be saved alongside it (click the *Preview* button or press *Cmd+Shift+K* to preview the HTML file). 

```{r}
library(ggplot2)
library(gridExtra)
library(phylogeo)

data(mountainsoil)

```



The preview shows you a rendered HTML copy of the contents of the editor. Consequently, unlike *Knit*, *Preview* does not run any R code chunks. Instead, the output of the chunk when it was last run in the editor is displayed.






