---
title: "siamcat_MND_MRD"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


##### ATTEMPT WITH MND ######

```{r}

#install.packages("randomForest")

library(SIAMCAT)
library(RColorBrewer)
library(pROC)
library(tidyverse)
library(curatedMetagenomicData)
library(ggpubr)

library(ExperimentHub)
library(curatedMetagenomicData)
suppressPackageStartupMessages(library(ExperimentHub))
library(phyloseq)
library(microbiome)




eh = ExperimentHub()
myquery = query(eh, "curatedMetagenomicData")

# Get the stool samples
esl <- curatedMetagenomicData("*metaphlan_bugs_list.stool*", dryrun = FALSE)

# merge
eset <- mergeData(esl)

# the relab=FALSE is essential as it Absolute Raw Count Data
pseq = ExpressionSet2phyloseq(eset,relab=FALSE ,simplify = TRUE, phylogenetictree = TRUE)
orig_pseq <- pseq


# Create mnd variable
#pseq <- orig_pseq
# grabs the nationality from phyloseq
c.mnd.country <- get_variable(pseq, "country" )
table(c.mnd.country)

# this collapses the two varialbes into a new variable AB in this case. The command for multiple changes is fct_collapse(x, AB = c("A","B"), DE = c("D","E"))
# Not used BGD Bangladesh , CAN Canada, CHN, China. DNK denmark, FJI Figi, HUN Hungry, ISL isle of man , MDG, MNG mongolia, PER Iran, RUS RUSSIA, 

#Adjsuted to match exactly the ATLAS data set

#central Europe = Belgium, Demark, Netherlands, Germany
#southern Europe = France, Italy, Serbia, Spain
#UKIE
#Scandinavia = Norway, Sweden Finland
#Eastern Europe = Poland 


c.mnd.atlas <- fct_collapse(c.mnd.country, Scandinavia = c("NOR","SWE","FIN") ,CentralEurope = c("DNK","DEU","NLD"), SouthEurope = c("ITA","ESP","FRA"), US = "USA", UKIE = "GBR" )

#drops unused levels
c.mnd.atlas <- droplevels(c.mnd.atlas,c("ISL","HUN","EST", "SVK", "LUX", "CAN", "AUT", "BGD","FJI","CHN","MDG","MNG","PER","RUS","TZA", "KAZ")) 

#c.mnd.atlas <- subset_samples(c.mnd.atlas, !is.na("country") )
#Error in access(object, "sam_data", errorIfNULL) : sam_data slot is empty.

sample_data(pseq)$nationality = c.mnd.atlas


#mnd.country <- fct_collapse(mnd.country, LOW = c("NOR","SWE","FIN","DNK","LUX") , MEDIUM = c("ITA","AUT","DEU","ESP","FRA","NLD"), HIGH = c("GBR","USA","CAN") )

#mnd.country <- fct_collapse(mnd.country, LOW = c("Scandinavia","EasternEurope") , MEDIUM = c("SouthEurope","CentralEurope", "UKIE"), HIGH = c("US") )

# reorder
# NOTE THERE IS NO EASTERN EUROPE
c.mnd.nationality <- fct_collapse(c.mnd.atlas, LOW = c("Scandinavia") , MEDIUM = c("SouthEurope","CentralEurope", "UKIE" ), HIGH = c("US") )

c.mnd.nationality <- factor(c.mnd.nationality, levels = (c("LOW", "MEDIUM", "HIGH")))
levels(c.mnd.nationality)

# creates a new variable in the phyloseq called mnd
sample_data(pseq)$mnd = c.mnd.nationality

# checks that it has worked.
get_variable(pseq, "mnd")
table(get_variable(pseq, "mnd"))





```


```{r}

#pseq.subset <- subset_samples(pseq, !is.na(mnd) & is.na(disease) | disease == "healthy" & age_category != "child" & age_category != "newborn" & age_category != "schoolage" )

```



# pseq Preprocessing
Transformed to relative abundance, creating the new pseq.r object, which is then filtered such that only OTUs with a mean greater than 10^-5 are kept.

```{r}
any(taxa_sums(pseq) == 0)
# FALSE if OTUs are found in all samples and we don't need to prune the taxa. If the answer is TRUE, some OTUs are not present in any samples. This is usually the case when data is subset to remove some samples. OTUs unique to those sample are not removed along with the samples. Therefore, it is important to check this everytime the phyloseq object is filtered for samples using `subset_samples` function.
head(meta(pseq))

pseq.1 <- prune_taxa(taxa_sums(pseq) > 0, pseq)

# check again if any OTUs are not present in any samples
any(taxa_sums(pseq.1) == 0)


ntaxa(pseq)
ntaxa(pseq.1)

ntaxa(pseq) - ntaxa(pseq.1)
 

head(meta(pseq.1))

```

```{r}

summarize_phyloseq(pseq.1)

```


```{r}
pseq.1@otu_table [1:3, 1:3]

```


```{r}

#pseq.r  = transform_sample_counts(pseq, function(x) x / sum(x) )
#pseq.pr = filter_taxa(pseq.r, function(x) var(x) > 1e-5, TRUE)

```


```{r}

pseq.c <- core(pseq, detection = 0.1/100, prevalence = 50/100)

#pseq.subset <- subset_samples(pseq.c, !is.na(mnd) & disease == "healthy" & age_category != "child" & age_category != "newborn" & age_category != "schoolage" & antibiotics_current_use != "yes")

pseq.subset <- subset_samples(pseq.c, disease == "healthy" & age_category != "child" & age_category != "newborn" & age_category != "schoolage" )

pseq.c <- microbiome::transform(pseq.subset, "compositional")


```



The taxonomic abundance profiles and the sample meatadata for the example dataset are available through the curatedMetagenomicsData R package.

clean the data a bit

```{r}

table(pseq@sam_data[["country"]])

table(pseq@sam_data[["mnd"]])

```





### Creating Siamcat Object ###
We have a phyloseq object, therefore:
Metadata = sam_data
Features = otu_table


The siamcat object is constructed using the siamcat() function. There are two ways to initialize it:

Features: You can provide a feature matrix, data.frame, or otu_table to the function (together with label and metadata information):

siamcat <- siamcat(feat=feat, label=label, meta=meta)
phyloseq: The alternative is to create a siamcat object directly out of a phyloseq object:

siamcat <- siamcat(phyloseq=phyloseq, label=label)


```{r}

feat=otu_table(pseq.c)
meta=sample_data(pseq.c)

#Two ways to create siamcat object:

label <- create.label(meta=sample_data(pseq.c),
    label = "mnd",
    case = c("HIGH", "MEDIUM"), control = "LOW")


#FRIST WAY
# siamcat <- siamcat(phyloseq=phyloseq, label=label)
siamcat <- siamcat(phyloseq=pseq.c, label=label)

#SECOND WAY:
# siamcat <- siamcat(feat=feat, label=label, meta=meta)
# siamcat <- siamcat(feat=otu_table, label=label, meta=sample_data(pseq))
# Error in validObject(.Object) : invalid class “phyloseq” object: 
 # An otu_table is required for most analysis / graphics in the phyloseq-package



```






The input data should be organized in the same way for every version of SIAMCAT. All files are in tab-separated column format


Label data: First row is expected to be #BINARY:1=[label for cases];-1=[label for controls]
Second row should contain the sample identifiers as tab-separated list (consistent with feature and metadata).
Third row is expected to contain the actual class labels (tab-separated), e.g. 1 for each case and -1 for each
control.
Note: Labels can take other numeric values (but not characters or strings); importantly, the label for cases has to
be greater than the one for controls.


Feature matrix: features (in rows) x samples (in columns)
First row should contain sample labels (consistent with label data), while the first column should contain feature
labels (e.g. taxonomic identifiers). The remaining entries are expected to be real values >= 0 that quantify the
abundance of each feature in each sample.


Metadata (optional): samples (in rows) x metadata (in columns)
Metadata needs to be converted to numerical values by the user (This is necessary for heatmap displays)!






```{r}

#Accessing and assigning slots
#Each slot in siamcat can be accessed by typing
  #slot_name(siamcat)

#e.g. for the eval_data slot you can types
  #eval_data(siamcat)
#There is one notable exception: the phyloseq slot has to be accessed with physeq(siamcat) due to technical reasons.
physeq(siamcat) 


#Slots will be filled during the SIAMCAT workflow by the package’s functions. However, if for any reason a slot needs to be assigned outside of the workflow, the following formula can be used:
    #slot_name(siamcat) <- object_to_assign

#e.g. to assign a new_label object to the label slot:
    #label(siamcat) <- new_label
#Please note that this may lead to unforeseen consequences…

```

```{r}

#There are two slots that have slots inside of them. First, the model_list slot has a models slot that contains the actual list of mlr models -can be accessed via models(siamcat)- and model.type which is a character with the name of the method used to train the model: model_type(siamcat).

#The phyloseq slot has a complex structure. However, unless the phyloseq object is created outside of the SIAMCAT workflow, only two slots of phyloseq slot will be occupied: the otu_table slot containing the features table and the sam_data slot containing metadata information. Both can be accessed by typing either features(siamcat) or meta(siamcat).

#Additional slots inside the phyloseq slots do not have dedicated accessors, but can easily be reached once the phyloseq object is exported from the siamcat object:

phyloseq <- physeq(siamcat)
otu_tab <- otu_table(phyloseq)
head(otu_tab)

```

Data can also be sub-selected based on the available meta-data. For example, if we want to exclude patients that are too young or too old for the question of interest, we can do so easily with:
```{r}

#siamcat <- select.samples(
#    siamcat,
#    filter = 'age',
#    allowed.set = NULL,
#    allowed.range = c(20, 90),
#    verbose = 2
#)

```




## Feature Filtering
Now, we can filter feature with low overall abundance and prevalence.
Since we have quite a lot of microbial species in the dataset at the moment, we can perform unsupervised feature selection using the function filter.features.
```{r}

siamcat <- filter.features(siamcat, cutoff=1e-04, filter.method = 'abundance')

siamcat <- filter.features(siamcat, cutoff=0.05,
                          filter.method='prevalence',
                          feature.type = 'filtered')

```

A few information about the siamcat object can be accessed with the show function from phyloseq (SIAMCAT builds on the phyloseq data structure):

```{r}

show(siamcat)

```






#### Association Testing ####
Checks for associations of single species with disease suing nonparametric Wilcoxon test.

Associations between microbial species and the label can be tested with the check.associations function. The function computes for each species the significance using a non-parametric Wilcoxon test and different effect sizes for the association (e.g. AUC or fold change).

The check.assocation function calculates the significance of enrichment and metrics of association (such as generalized fold change and single-feautre AUROC).

The function produces a pdf file as output, since the plot is optimized for a landscape DIN-A4 layout, but can also used to plot on an active graphic device, e.g. in RStudio. The resulting plot then looks like that:
```{r}

library(progress)
library(pROC)
library(RColorBrewer)

siamcat <- check.associations(
    siamcat,
    sort.by = 'fc',
    fn.plot = 'assocMND_healthy.pdf',
    alpha = 0.05,
    mult.corr = "fdr",
    detect.lim = 10 ^-6,
    plot.type = "quantile.box",
    panels = c("fc", "prevalence", "auroc"))

```

Confounders are checked with the function check.confounders, which produces a plot for each possible confounder in the metadata and diverts the output into a pdf-file.

```{r}

check.confounders(siamcat,
    fn.plot = 'conf_checkMND_healthy.pdf', meta.in = c("mnd", "gender", "age_category"))

```


##### Model Building ####

One strength of SIAMCAT is the versatile but easy-to-use interface for the construction of machine learning models on the basis of microbial species. SIAMCAT contains functions for data normalization, splitting the data into cross-validation folds, training the model, and making predictions based on cross-validation instances and the trained models.


#### Data Normalization

Data normalization is performed with the normalize.features function. Here, we use the log.unit method, but several other methods and customization options are available (please check the documentation).

Getting all data on the same scale: if the scales for different features are wildly different, this can have a knock-on effect on your ability to learn (depending on what methods you're using to do it). Ensuring standardised feature values implicitly weights all features equally in their representation.

```{r}

siamcat <- normalize.features(
    siamcat,
    norm.method = "log.unit",
    norm.param = list(
        log.n0 = 1e-06,
        n.p = 2,
        norm.margin = 1
    )
)

```




#### Perpare Cross-Validation

https://www.youtube.com/watch?v=fSytzGwwBVw

Cross-Validation is basically how we decide which machine learning method would be best for our data set. CV allows us to compare different machine learning methods and get a sense of how well they will work in practice.

Examples of machine learning methods:
1. Logistic Regression
2. K-nearest neighbours
3. Support Vector Machines (SVM) ...etc.

Need to do 2 things with collected data:
1. Estimate parameters for machine learning methods. (training algorithm)
2. Evaluate how well the machine learning method works. (testing algorithm)

Terrible approach to use ALL the data to train algorithm bc wouldn't have any data left to test algorithm. - Data separated into blocks, different block combinations used to train model.


Preparation of the cross-validation fold is a crucial step in machine learning. SIAMCAT greatly simplifies the set-up of cross-validation schemes, including stratification of samples or keeping samples inseperable based on metadata. For this small example, we choose a twice-repeated 5-fold (no. of blocks) cross-validation scheme. The data-split will be saved in the data_split slot of the siamcat object.
```{r}

siamcat <-  create.data.split(
    siamcat,
    num.folds = 5,
    num.resample = 2
)

```



#### Model Training

The actual model training is performed using the function train.model. Again, multiple options for customization are available, ranging from the machine learning method to the measure for model selection or customizable parameter set for hyperparameter tuning.
```{r}

#library(randomForest)

siamcat <- train.model(
    siamcat,
    method = "lasso"
)


#siamcat <- train.model(
#    siamcat,
#    method = "enet"
#)


#siamcat <- train.model(
#    siamcat,
#    method = "ridge"
#)


#siamcat <- train.model(
#    siamcat,
#    method = "lasso_ll"
#)


#siamcat <- train.model(
#    siamcat,
#    method = "ridge_ll"
#)


#siamcat <- train.model(
#    siamcat,
#    method = "randomForest"
#)


```

The models are saved in the model_list slot of the siamcat object. The model building is performed using the mlr R package. All models can easily be accessed.
```{r}

# get information about the model type
model_type(siamcat)

# access the models
models <- models(siamcat)
models[[1]]

```


#### Make Predicitions

Using the data-split and the models trained in previous step, we can use the function make.predictions in order to apply the models on the test instances in the data-split. The predictions will be saved in the pred_matrix slot of the siamcat object.

```{r}

siamcat <- make.predictions(siamcat)
pred_matrix <- pred_matrix(siamcat)

head(pred_matrix)

```



#### Model Evaluation and Interpretation ####

In the final part, we want to find out how well the model performed and which microbial species had been selected in the model. In order to do so, we first calculate how well the predictions fit the real data using the function evaluate.predictions. This function calculates the Area Under the Receiver Operating Characteristic (ROC) Curve (AU-ROC) and the Precision Recall (PR) Curve for each resampled cross-validation run.

```{r}

siamcat <-  evaluate.predictions(siamcat)

```



#### Evaluation Plot

To plot the results of the evaluation, we can use the function model.evaluation.plot, which produces a pdf-file showing the ROC and PR Curves for the different resamples runs as well as the mean ROC and PR Curve.

```{r}

model.evaluation.plot(siamcat)

```



#### Interpretation Plot
Displays importance of individual features in the classification model

The final plot produced by SIAMCAT is the model interpretation plot, created by the model.interpretation.plot function. The plot shows for the top selected features the:

  - model weights (and how robust they are, i.e. in what proportion of models have they been incorporated) as a barplot

  - a heatmap with the z-scores or fold changes for the top selected features

  - a boxplot showing the proportions of weight per model which is captured by the top selected features.

Additionally, the distribution of metadata is shown in a heatmap below.

The function again produces a pdf-file optimized for a landscape DIN-A4 plotting region.

```{r}

model.interpretation.plot(
    siamcat, 
    fn.plot = 'interpretationMND_healthy.pdf',
    color.scheme = "BrBG",
    consens.thres = 0.5,
    heatmap.type = "zscore",
    limits = c(-3, 3), detect.lim = 1e-06,
    max.show = 50, prompt=TRUE, verbose = 1)


model.interpretation.plot(
    siamcat, 
    fn.plot = 'interpretationfcMND.pdf',
   color.scheme = "BrBG",
    consens.thres = 0.5,
    heatmap.type = "fc",
    limits = c(-3, 3), detect.lim = 1e-06,
    max.show = 50, prompt=TRUE, verbose = 3)

```
















#### Confounder Analysis

We can also check the supplied meta-variables for potential confounding.

```{r}

#check.confounders(siamcat, fn.plot = './confoundersMND.pdf')

```




we have more samples than individual subjects:
```{r}

#print(length(unique(pseq$subjectID))) 
#print(nrow(pseq))


#print(length(unique(pseq@sam_data[["subjectID"]])))
#6784 total rows


```

Some subjects (but not all) had been sampled multiple times. Therefore, we want to remove repeated samplings for the same subject, since the samples would otherwise not be indepdenent from another.

The visit number is encoded in the sampleID. Therefore, we can use this information to extract when the samples have been taken and use only the first visit for each subject.

```{r}

#pseq.s<- pseq %>%
  #select(sampleID, subjectID, study_condition, disease_subtype, 
#disease, age, country, number_reads, median_read_length, BMI, mnd) %>%
#  mutate(visit=str_extract(sampleID, '_[0-9]+$')) %>%
#  mutate(visit=str_remove(visit, '_')) %>% 
#  mutate(visit=as.numeric(visit)) %>%
 # mutate(visit=case_when(is.na(visit)~0, TRUE~visit)) %>%
  #group_by(subjectID) %>% 
  #filter(visit==min(visit)) %>%
  #ungroup() %>%
  #mutate(Sample_ID=sampleID) %>%
  #mutate(Group=case_when(mnd=='LOW'~'CTR'))
  

# mutate(Group=case_when(disease=='healthy'~'CTR')
```







